{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c69fd7cb-9072-46e4-8f1e-a840600ed0cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "418 747\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import glob, os, vcf, warnings, shutil, subprocess, re, sys\n",
    "from Bio import Seq, SeqIO\n",
    "warnings.filterwarnings('ignore')\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats as st\n",
    "import scipy\n",
    "from Bio import Seq, SeqIO, Entrez\n",
    "from dna_features_viewer import BiopythonTranslator\n",
    "# import panel as pn\n",
    "# pn.extension('ipywidgets')\n",
    "# pn.extension()\n",
    "\n",
    "h37Rv_path = \"/n/data1/hms/dbmi/farhat/Sanjana/H37Rv\"\n",
    "h37Rv_seq = SeqIO.read(os.path.join(h37Rv_path, \"GCF_000195955.2_ASM19595v2_genomic.gbff\"), \"genbank\")\n",
    "h37Rv_genes = pd.read_csv(os.path.join(h37Rv_path, \"mycobrowser_h37rv_genes_v4.csv\"))\n",
    "h37Rv_regions = pd.read_csv(os.path.join(h37Rv_path, \"mycobrowser_h37rv_v4.csv\"))\n",
    "\n",
    "h37Rv_coords = pd.read_csv(os.path.join(h37Rv_path, \"h37Rv_coords_to_gene.csv\"))\n",
    "h37Rv_coords_dict = dict(zip(h37Rv_coords[\"pos\"].values, h37Rv_coords[\"region\"].values))\n",
    "\n",
    "# keep only the 727 patients with high quality WGS and matched patients\n",
    "df_TRUST_pids_samples = pd.read_csv(\"/home/sak0914/MtbQuantCNN/analysis/TRUST/processed_data/combined_patient_WGS_data.csv\").dropna(subset=['pid', 'Lineage']).reset_index(drop=True)\n",
    "print(df_TRUST_pids_samples.pid.nunique(), df_TRUST_pids_samples.SampleID.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef2e40a5-0419-4ba2-b500-5f15c3f496d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8d28fe72-b4f1-434d-a6da-e626b0a845fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_TRUST_pids_samples['Lineage'] = df_TRUST_pids_samples['Lineage'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0b8deead-abe6-4ffa-b252-d054d8631caf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mixed_samples = df_TRUST_pids_samples.query(\"F2 > 0.03\").SampleID.unique()\n",
    "len(mixed_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "538b4b33-0c6e-4ca3-aa72-14aeeb0822b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_samples = list(set(mixed_samples) - set(os.listdir(output_dir)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f216ae4c-c39b-470e-95b7-aac0b31e86d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30, 30)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_isolates = pd.DataFrame(columns = ['SampleID', 'Percents'])\n",
    "i = 0\n",
    "\n",
    "for sample in os.listdir(output_dir):\n",
    "\n",
    "    combined_percents_string = []\n",
    "    \n",
    "    for fName in glob.glob(f\"{output_dir}/{sample}/*_strain.reads\"):\n",
    "        combined_percents_string.append(os.path.basename(fName).split('_strain.reads')[0])\n",
    "\n",
    "    if len(combined_percents_string) >= 0:\n",
    "        df_isolates.loc[i, :] = [sample, ','.join(combined_percents_string)]\n",
    "        i += 1\n",
    "\n",
    "len(df_isolates), len(os.listdir(output_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "4a8f6585-7bf0-40fc-908c-98e1826a0ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_isolates.to_csv(\"./isolates_to_run.tsv\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0898628d-5b99-4e5c-bd2a-b5a6fb212512",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pid</th>\n",
       "      <th>Original_ID</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>SampleLib</th>\n",
       "      <th>Pacbio</th>\n",
       "      <th>Comment</th>\n",
       "      <th>quality</th>\n",
       "      <th>Cov_Any_Mean</th>\n",
       "      <th>Cov_Unam_Perc</th>\n",
       "      <th>Perc_Reads_Mapped</th>\n",
       "      <th>...</th>\n",
       "      <th>fib4</th>\n",
       "      <th>prevtb_outcome</th>\n",
       "      <th>total_adherence</th>\n",
       "      <th>adherence_12week</th>\n",
       "      <th>first_dots_date</th>\n",
       "      <th>last_dots_date</th>\n",
       "      <th>F2</th>\n",
       "      <th>Coll2014</th>\n",
       "      <th>Freschi2020</th>\n",
       "      <th>Lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>T0087</td>\n",
       "      <td>S0087-01</td>\n",
       "      <td>MFS-353</td>\n",
       "      <td>MFS-353_lib59077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>422.31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.9897</td>\n",
       "      <td>...</td>\n",
       "      <td>0.614105</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.983607</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>178</td>\n",
       "      <td>0.415768</td>\n",
       "      <td>4.1.2.1,4.8</td>\n",
       "      <td>4.2.1.1.1.1.1.1.i1,4.1.i1.1.1.1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>T0072</td>\n",
       "      <td>S0072-01</td>\n",
       "      <td>MFS-350</td>\n",
       "      <td>MFS-350_lib59074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>346.84</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.9899</td>\n",
       "      <td>...</td>\n",
       "      <td>0.285549</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.819549</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>1</td>\n",
       "      <td>189</td>\n",
       "      <td>0.408194</td>\n",
       "      <td>4.1.2.1,4.8</td>\n",
       "      <td>4.2.1.1.1.1.1.1.i1,4.1.i1.1.1.1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 7101 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       pid Original_ID SampleID         SampleLib  Pacbio Comment  quality  \\\n",
       "379  T0087    S0087-01  MFS-353  MFS-353_lib59077     NaN     NaN      1.0   \n",
       "400  T0072    S0072-01  MFS-350  MFS-350_lib59074     NaN     NaN      1.0   \n",
       "\n",
       "     Cov_Any_Mean  Cov_Unam_Perc  Perc_Reads_Mapped  ...      fib4  \\\n",
       "379        422.31            1.0             0.9897  ...  0.614105   \n",
       "400        346.84            1.0             0.9899  ...  0.285549   \n",
       "\n",
       "     prevtb_outcome  total_adherence  adherence_12week  first_dots_date  \\\n",
       "379             2.0         0.983607          1.000000                1   \n",
       "400             NaN         0.819549          0.833333                1   \n",
       "\n",
       "     last_dots_date        F2     Coll2014                      Freschi2020  \\\n",
       "379             178  0.415768  4.1.2.1,4.8  4.2.1.1.1.1.1.1.i1,4.1.i1.1.1.1   \n",
       "400             189  0.408194  4.1.2.1,4.8  4.2.1.1.1.1.1.1.i1,4.1.i1.1.1.1   \n",
       "\n",
       "    Lineage  \n",
       "379       4  \n",
       "400       4  \n",
       "\n",
       "[2 rows x 7101 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_TRUST_pids_samples.query(\"SampleID in ['MFS-350', 'MFS-353']\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58fe320e-8ccf-4a2b-96f2-9bf805ee05d4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7503ee1a-8d09-4ac8-b9d4-1f60e7811075",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MFS-134 0.0 0.87 0.13 \n",
      "MFS-145 0.0 0.87 0.13 \n",
      "MFS-198 0.0 0.85 0.15 \n",
      "MFS-270 0.0 0.79 0.21 \n",
      "MFS-313 0.0 0.88 0.12 \n",
      "MFS-317 0.0 0.71 0.29 \n",
      "MFS-336 0.0 0.52 0.48 \n",
      "MFS-350 1.0 1.0\n",
      "MFS-353 1.0 1.0\n",
      "MFS-377 0.0 0.87 0.13 \n",
      "MFS-39 0.0 0.86 0.14 \n",
      "MFS-450 0.0 0.83 0.17 \n",
      "MFS-456 0.0 0.86 0.14 \n",
      "MFS-485 0.0 0.86 0.14 \n",
      "MFS-486 0.0 0.85 0.15 \n",
      "MFS-577 0.0 0.85 0.15 \n",
      "MFS-608 0.0 0.85 0.15 \n",
      "MFS-630 0.0 0.85 0.15 \n",
      "MFS-631 0.0 0.74 0.26 \n",
      "MFS-696 0.0 0.84 0.16 \n",
      "MFS-835 0.0 0.86 0.14 \n",
      "MFS-84 0.0 0.82 0.18 \n",
      "MFS-85 0.0 0.76 0.24 \n"
     ]
    }
   ],
   "source": [
    "for sample in os.listdir(output_dir):\n",
    "    try:\n",
    "        result = pd.read_csv(f\"{output_dir}/{sample}/result.txt\", sep='\\t', comment='I')\n",
    "        print(sample, result['p-value'].values[0], result['proportions'].values[0])\n",
    "    except:\n",
    "        print(f\"{sample} not yet finished\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9c59f56e-63d0-4d6d-9afd-4efc6db3440d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_variants = pd.read_csv(f\"../processed_data/phase_variation_regions/indels_HT.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b1ad18ca-bc6c-4093-a149-ff94e797d64b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF</th>\n",
       "      <th>F2</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>Coll2014</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFS-336</td>\n",
       "      <td>2536628</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>0.480952</td>\n",
       "      <td>0.458952</td>\n",
       "      <td>2,4</td>\n",
       "      <td>4.1.1.1,2.2.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MFS-313</td>\n",
       "      <td>2536628</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>0.472527</td>\n",
       "      <td>0.436810</td>\n",
       "      <td>2,4</td>\n",
       "      <td>4.3.2.1,2.2.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MFS-313</td>\n",
       "      <td>854252</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.436810</td>\n",
       "      <td>2,4</td>\n",
       "      <td>4.3.2.1,2.2.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MFS-317</td>\n",
       "      <td>854252</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>0.626062</td>\n",
       "      <td>0.369685</td>\n",
       "      <td>2,4</td>\n",
       "      <td>4.3.2.1,2.2.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MFS-317</td>\n",
       "      <td>1572680</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>0.632530</td>\n",
       "      <td>0.369685</td>\n",
       "      <td>2,4</td>\n",
       "      <td>4.3.2.1,2.2.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>MFS-317</td>\n",
       "      <td>2536628</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>0.616541</td>\n",
       "      <td>0.369685</td>\n",
       "      <td>2,4</td>\n",
       "      <td>4.3.2.1,2.2.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MFS-631</td>\n",
       "      <td>2536628</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>0.732558</td>\n",
       "      <td>0.314259</td>\n",
       "      <td>2,3</td>\n",
       "      <td>2.2.1,3.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MFS-85</td>\n",
       "      <td>854252</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>0.729299</td>\n",
       "      <td>0.259418</td>\n",
       "      <td>2,4</td>\n",
       "      <td>2.2.1,4.1.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MFS-85</td>\n",
       "      <td>2536628</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>0.646226</td>\n",
       "      <td>0.259418</td>\n",
       "      <td>2,4</td>\n",
       "      <td>2.2.1,4.1.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MFS-630</td>\n",
       "      <td>36470</td>\n",
       "      <td>GCC</td>\n",
       "      <td>G</td>\n",
       "      <td>0.177083</td>\n",
       "      <td>0.147559</td>\n",
       "      <td>2</td>\n",
       "      <td>2.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MFS-134</td>\n",
       "      <td>4323354</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>0.566434</td>\n",
       "      <td>0.138670</td>\n",
       "      <td>2</td>\n",
       "      <td>2.2.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>MFS-456</td>\n",
       "      <td>4358979</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>0.098958</td>\n",
       "      <td>0.130840</td>\n",
       "      <td>2</td>\n",
       "      <td>2.2.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MFS-145</td>\n",
       "      <td>2536628</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>0.113139</td>\n",
       "      <td>0.122113</td>\n",
       "      <td>4</td>\n",
       "      <td>4.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MFS-145</td>\n",
       "      <td>4323354</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>0.109705</td>\n",
       "      <td>0.122113</td>\n",
       "      <td>4</td>\n",
       "      <td>4.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>MFS-696</td>\n",
       "      <td>1572680</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>0.088670</td>\n",
       "      <td>0.108637</td>\n",
       "      <td>2</td>\n",
       "      <td>2.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MFS-198</td>\n",
       "      <td>2536628</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>0.069652</td>\n",
       "      <td>0.075920</td>\n",
       "      <td>4</td>\n",
       "      <td>4.1.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>MFS-198</td>\n",
       "      <td>4358186</td>\n",
       "      <td>G</td>\n",
       "      <td>GT</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.075920</td>\n",
       "      <td>4</td>\n",
       "      <td>4.1.1.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SampleID      POS  REF ALT        AF        F2 Lineage         Coll2014\n",
       "1   MFS-336  2536628    G  GA  0.480952  0.458952     2,4  4.1.1.1,2.2.1.1\n",
       "29  MFS-313  2536628    G  GA  0.472527  0.436810     2,4  4.3.2.1,2.2.1.1\n",
       "28  MFS-313   854252   GC   G  0.404844  0.436810     2,4  4.3.2.1,2.2.1.1\n",
       "30  MFS-317   854252   GC   G  0.626062  0.369685     2,4  4.3.2.1,2.2.1.1\n",
       "31  MFS-317  1572680    G  GC  0.632530  0.369685     2,4  4.3.2.1,2.2.1.1\n",
       "32  MFS-317  2536628    G  GA  0.616541  0.369685     2,4  4.3.2.1,2.2.1.1\n",
       "19  MFS-631  2536628    G  GA  0.732558  0.314259     2,3      2.2.1,3.1.1\n",
       "11   MFS-85   854252   GC   G  0.729299  0.259418     2,4    2.2.1,4.1.1.3\n",
       "12   MFS-85  2536628    G  GA  0.646226  0.259418     2,4    2.2.1,4.1.1.3\n",
       "17  MFS-630    36470  GCC   G  0.177083  0.147559       2            2.2.1\n",
       "22  MFS-134  4323354    C  CG  0.566434  0.138670       2          2.2.1.1\n",
       "40  MFS-456  4358979    C  CG  0.098958  0.130840       2          2.2.1.1\n",
       "24  MFS-145  2536628    G  GA  0.113139  0.122113       4              4.8\n",
       "25  MFS-145  4323354    C  CG  0.109705  0.122113       4              4.8\n",
       "47  MFS-696  1572680    G  GC  0.088670  0.108637       2            2.2.1\n",
       "34  MFS-198  2536628    G  GA  0.069652  0.075920       4          4.1.1.3\n",
       "35  MFS-198  4358186    G  GT  0.058824  0.075920       4          4.1.1.3"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variants.merge(df_isolates, on='SampleID')[['SampleID', 'POS', 'REF', 'ALT', 'AF', 'F2', 'Lineage', 'Coll2014']].query(\"AF >= 0.05 & AF <= 0.75\").sort_values(\"F2\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d98e998-342e-47ad-8d31-edcc840971d1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
