{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b3946e2b-1325-4a1b-94a8-7d3202f2f026",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os, vcf, sys, itertools, shutil, subprocess, glob, yaml\n",
    "from Bio import Entrez, Seq, SeqIO\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "H37Rv_regions = pd.read_csv(\"./references/ref_genome/mycobrowser_h37rv_v4.csv\")\n",
    "isolate_metadata = pd.read_csv(\"/n/data1/hms/dbmi/farhat/rollingDB/metadata/isolate_metadata.csv\")\n",
    "\n",
    "df_WGS = pd.read_csv(\"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/Illumina_culture_WGS_summary.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ec9c179d-86b0-42ae-be57-39f31e3b073a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "206 179\n"
     ]
    }
   ],
   "source": [
    "DS_samples = os.listdir(\"/n/data1/hms/dbmi/farhat/aryan/TRUST/FreeBayesVCFs/AF_replaced_vcfs/DS\")\n",
    "\n",
    "# not sure if we should remove the A and B samples\n",
    "# DS_samples = [fName.split('_dedup')[0].replace('_', '-').replace('A', '').replace('B', '') for fName in DS_samples]\n",
    "DS_samples = [fName.split('_')[0].replace('S0', 'T0') for fName in DS_samples]\n",
    "print(len(DS_samples), len(np.unique(DS_samples)))\n",
    "\n",
    "DS_samples = np.unique(DS_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1972ffbf-f4f8-458e-95b1-e4b40105cf7e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trust_report_fNames = glob.glob(\"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS_metadata_reports/*\")\n",
    "len(trust_report_fNames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3bd9e94e-a17d-4bbc-ad09-c7cc82dd2b1d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_trust_WGS_metadata = pd.read_csv(\"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/clinical_data/combined_patient_WGS_data.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "5b90b3f7-e877-47a8-bd0b-4112331f04f4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "copy_samples = df_trust_WGS_metadata.query(\"pid in @DS_samples\")['Original_ID'].str.replace('_', '-').unique()\n",
    "\n",
    "def normalize_patient_sample_prefixes(samples_lst):\n",
    "    '''\n",
    "    Add S padded with 0s to make the full name S#### so we can match the samples to patients.\n",
    "    '''\n",
    "    \n",
    "    normalized_sample_names = []\n",
    "    \n",
    "    for sample in samples_lst:\n",
    "        \n",
    "        # I think these are TOTAL samples. Just return the name as is because there's no longitudinal component, suggesting it's not a TRUST sample\n",
    "        if '-' not in sample:\n",
    "            normalized_sample_names.append(sample)\n",
    "        \n",
    "        else:\n",
    "            prefix, suffix = sample.split(\"-\", 1)\n",
    "            if prefix.startswith(\"S\"):\n",
    "                # already starts with S, just ensure 4 digits after S\n",
    "                num_part = prefix[1:].zfill(4)  \n",
    "                new_prefix = f\"S{num_part}\"\n",
    "            else:\n",
    "                # no S, add it and zero-pad to 4 digits\n",
    "                new_prefix = f\"S{prefix.zfill(4)}\"\n",
    "            normalized_sample_names.append(f\"{new_prefix}-{suffix}\")\n",
    "\n",
    "    return normalized_sample_names\n",
    "\n",
    "normalized_sample_names = normalize_patient_sample_prefixes(copy_samples)\n",
    "\n",
    "df_Aryan_samples = pd.DataFrame({'Name': copy_samples, 'NormName': normalized_sample_names})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "2eceb6f3-3c52-42ed-9c96-8b7a37183f10",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>NormName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S0157-01</td>\n",
       "      <td>S0157-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>157-01</td>\n",
       "      <td>S0157-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>170-01</td>\n",
       "      <td>S0170-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>S0170-01</td>\n",
       "      <td>S0170-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Name  NormName\n",
       "5    S0157-01  S0157-01\n",
       "6      157-01  S0157-01\n",
       "154    170-01  S0170-01\n",
       "155  S0170-01  S0170-01"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Aryan_samples.iloc[df_Aryan_samples.index.values[df_Aryan_samples.duplicated(subset='NormName', keep=False)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "df0ed728-c173-4860-b200-91483d862986",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "in_dir = \"/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Directsputum\"\n",
    "out_dir = \"/n/scratch/users/s/sak0914/TRUST_sputum\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "922ba03c-1b39-4f29-ac4b-bbeac7bc5f5c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "542"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FQ_files = glob.glob(f\"{in_dir}/*/*.gz\")\n",
    "\n",
    "df_FQ = pd.DataFrame({'FQ': FQ_files})\n",
    "df_FQ['fName'] = [os.path.basename(fName) for fName in df_FQ['FQ']]\n",
    "df_FQ['dirName'] = [os.path.basename(os.path.dirname(fName)) for fName in df_FQ['FQ']]\n",
    "len(df_FQ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "693d1687-9026-4667-a618-5b060f4a8284",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1' '2']\n"
     ]
    }
   ],
   "source": [
    "df_FQ['Sample'] = [os.path.basename(fName).split('_S')[0].split('_R')[0].replace('_', '-') for fName in df_FQ['FQ']]\n",
    "\n",
    "# normalize names\n",
    "df_FQ['Sample'] = normalize_patient_sample_prefixes(df_FQ['Sample'])\n",
    "\n",
    "df_FQ['ReadNum'] = [os.path.basename(fName).split('_R')[1].split('_')[0].split('.')[0] for fName in df_FQ['FQ']]\n",
    "print(df_FQ.ReadNum.unique())\n",
    "\n",
    "df_FQ['ReadNum'] = df_FQ['ReadNum'].astype(int)\n",
    "df_FQ[['Original_ID', 'SampleNum']] = df_FQ['Sample'].str.split('-', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "98c15c90-ec9f-4326-91d9-a135fc1e336b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "388"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_FQ_to_run = df_FQ.drop_duplicates(subset=['Sample', 'ReadNum'], keep=False)\n",
    "len(df_FQ_to_run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "e3f80dc8-d139-4480-ae69-eb9d595f6f76",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FQ</th>\n",
       "      <th>fName</th>\n",
       "      <th>dirName</th>\n",
       "      <th>Sample</th>\n",
       "      <th>ReadNum</th>\n",
       "      <th>Original_ID</th>\n",
       "      <th>SampleNum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Direct...</td>\n",
       "      <td>S0066_01_R1.fastq.gz</td>\n",
       "      <td>20220509_TwistV1</td>\n",
       "      <td>S0066-01</td>\n",
       "      <td>1</td>\n",
       "      <td>S0066</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Direct...</td>\n",
       "      <td>S0066_01_R2.fastq.gz</td>\n",
       "      <td>20220509_TwistV1</td>\n",
       "      <td>S0066-01</td>\n",
       "      <td>2</td>\n",
       "      <td>S0066</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   FQ                 fName  \\\n",
       "24  /n/data1/hms/dbmi/farhat/fastq_db/TRUST_Direct...  S0066_01_R1.fastq.gz   \n",
       "25  /n/data1/hms/dbmi/farhat/fastq_db/TRUST_Direct...  S0066_01_R2.fastq.gz   \n",
       "\n",
       "             dirName    Sample  ReadNum Original_ID SampleNum  \n",
       "24  20220509_TwistV1  S0066-01        1       S0066        01  \n",
       "25  20220509_TwistV1  S0066-01        2       S0066        01  "
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_FQ.query(\"fName.str.contains('066')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "a12152a1-d1a8-4ea7-93d9-7a8ebefa92df",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "117"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_FQ_to_run.query(\"Sample in @normalized_sample_names\").Sample.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "39dacaa1-dccf-4067-8c2d-97c83b635836",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "234\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FQ</th>\n",
       "      <th>fName</th>\n",
       "      <th>dirName</th>\n",
       "      <th>Sample</th>\n",
       "      <th>ReadNum</th>\n",
       "      <th>Original_ID</th>\n",
       "      <th>SampleNum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Direct...</td>\n",
       "      <td>S0066_01_R1.fastq.gz</td>\n",
       "      <td>20220509_TwistV1</td>\n",
       "      <td>S0066-01</td>\n",
       "      <td>1</td>\n",
       "      <td>S0066</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Direct...</td>\n",
       "      <td>S0066_01_R2.fastq.gz</td>\n",
       "      <td>20220509_TwistV1</td>\n",
       "      <td>S0066-01</td>\n",
       "      <td>2</td>\n",
       "      <td>S0066</td>\n",
       "      <td>01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   FQ                 fName  \\\n",
       "24  /n/data1/hms/dbmi/farhat/fastq_db/TRUST_Direct...  S0066_01_R1.fastq.gz   \n",
       "25  /n/data1/hms/dbmi/farhat/fastq_db/TRUST_Direct...  S0066_01_R2.fastq.gz   \n",
       "\n",
       "             dirName    Sample  ReadNum Original_ID SampleNum  \n",
       "24  20220509_TwistV1  S0066-01        1       S0066        01  \n",
       "25  20220509_TwistV1  S0066-01        2       S0066        01  "
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_FQ_to_copy = df_FQ_to_run.query(\"Sample in @normalized_sample_names\")\n",
    "print(len(df_FQ_to_copy))\n",
    "\n",
    "df_FQ_to_copy.query(\"FQ.str.contains('066')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "214d5b58-2c9c-4928-a128-8044b0e5afc1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "77"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_samples = list(set(df_FQ_to_run.Sample) - set(df_FQ_to_run.query(\"Sample in @normalized_sample_names\").Sample))\n",
    "len(missing_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "85d3c63c-1061-456b-be96-0dad3eca03a1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['20220509_TwistV1', '20240819_TwistV2'], dtype=object)"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_FQ_to_copy.dirName.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "052ef154-cf35-4f7d-ac8a-0cbca39e39a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# for i, row in df_FQ_to_copy.iterrows():\n",
    "\n",
    "#     os.makedirs(f\"{out_dir}/{row['Sample']}\", exist_ok=True)\n",
    "    \n",
    "#     shutil.copy(row['FQ'], f\"{out_dir}/{row['Sample']}/{row['Sample']}_R{row['ReadNum']}.fastq.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "ef602a2a-b76c-4217-82d1-bfb480616379",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_samples = df_FQ_to_copy[['Sample', 'Sample']].drop_duplicates()\n",
    "df_samples['FQ'] = 0\n",
    "\n",
    "df_samples.to_csv(\"~/longitudinal_changes/Mtb_low_frequency_variants/TRUST_DS_isolates.tsv\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bd565c81-896a-4aa6-88bd-28fd565b802e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame({'A': 'S0326-01',\n",
    "                   'B': 'S0326-01',\n",
    "                   'C': 0\n",
    "                  }, index=[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3a954d21-64c2-4f50-97ad-3e19713bc2a7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df.to_csv(\"/home/sak0914/longitudinal_changes/Mtb_low_frequency_variants/TRUST_DS_isolates.tsv\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ecd0f7c7-18ec-4100-8584-0c821dca91ba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_kraken = pd.read_csv(\"/n/data1/hms/dbmi/farhat/Sanjana/TRUST_direct_sputum/S0326-01/S0326-01/kraken/kraken_report_standard_DB.txt\", sep='\\t', header=None,\n",
    "                        names=['Percent', 'CumulCount', 'Count', 'TaxLevel', 'Taxid', 'Name']\n",
    "                       )\n",
    "\n",
    "df_kraken['Name'] = df_kraken['Name'].str.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4b4bef52-0380-4bb2-baeb-351a3702510c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Percent</th>\n",
       "      <th>CumulCount</th>\n",
       "      <th>Count</th>\n",
       "      <th>TaxLevel</th>\n",
       "      <th>Taxid</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>53.03</td>\n",
       "      <td>379055</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>9605</td>\n",
       "      <td>Homo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>16.64</td>\n",
       "      <td>118902</td>\n",
       "      <td>98625</td>\n",
       "      <td>G</td>\n",
       "      <td>1763</td>\n",
       "      <td>Mycobacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>14.54</td>\n",
       "      <td>103925</td>\n",
       "      <td>63260</td>\n",
       "      <td>G</td>\n",
       "      <td>1301</td>\n",
       "      <td>Streptococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>9.83</td>\n",
       "      <td>70278</td>\n",
       "      <td>37791</td>\n",
       "      <td>G</td>\n",
       "      <td>1378</td>\n",
       "      <td>Gemella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>807</th>\n",
       "      <td>1.57</td>\n",
       "      <td>11236</td>\n",
       "      <td>184</td>\n",
       "      <td>G</td>\n",
       "      <td>32207</td>\n",
       "      <td>Rothia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>0.33</td>\n",
       "      <td>2375</td>\n",
       "      <td>1513</td>\n",
       "      <td>G</td>\n",
       "      <td>29465</td>\n",
       "      <td>Veillonella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813</th>\n",
       "      <td>0.12</td>\n",
       "      <td>848</td>\n",
       "      <td>359</td>\n",
       "      <td>G</td>\n",
       "      <td>1269</td>\n",
       "      <td>Micrococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>0.09</td>\n",
       "      <td>644</td>\n",
       "      <td>104</td>\n",
       "      <td>G</td>\n",
       "      <td>1716</td>\n",
       "      <td>Corynebacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>0.07</td>\n",
       "      <td>517</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>1386</td>\n",
       "      <td>Bacillus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>0.05</td>\n",
       "      <td>328</td>\n",
       "      <td>121</td>\n",
       "      <td>G</td>\n",
       "      <td>57493</td>\n",
       "      <td>Kocuria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>0.05</td>\n",
       "      <td>377</td>\n",
       "      <td>38</td>\n",
       "      <td>G</td>\n",
       "      <td>1866885</td>\n",
       "      <td>Mycolicibacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>0.04</td>\n",
       "      <td>266</td>\n",
       "      <td>14</td>\n",
       "      <td>G</td>\n",
       "      <td>1663</td>\n",
       "      <td>Arthrobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1037</th>\n",
       "      <td>0.04</td>\n",
       "      <td>266</td>\n",
       "      <td>4</td>\n",
       "      <td>G</td>\n",
       "      <td>1912216</td>\n",
       "      <td>Cutibacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>0.03</td>\n",
       "      <td>200</td>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>1883</td>\n",
       "      <td>Streptomyces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>869</th>\n",
       "      <td>0.02</td>\n",
       "      <td>120</td>\n",
       "      <td>102</td>\n",
       "      <td>G</td>\n",
       "      <td>33882</td>\n",
       "      <td>Microbacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>885</th>\n",
       "      <td>0.02</td>\n",
       "      <td>118</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>529881</td>\n",
       "      <td>Candidatus Aquiluna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>0.02</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>G</td>\n",
       "      <td>157920</td>\n",
       "      <td>Cellulosimicrobium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>0.01</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>1696</td>\n",
       "      <td>Brevibacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1185</th>\n",
       "      <td>0.01</td>\n",
       "      <td>85</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>1654</td>\n",
       "      <td>Actinomyces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1469</th>\n",
       "      <td>0.01</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>68</td>\n",
       "      <td>Lysobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1481</th>\n",
       "      <td>0.01</td>\n",
       "      <td>36</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>286</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1065</th>\n",
       "      <td>0.01</td>\n",
       "      <td>47</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>1839</td>\n",
       "      <td>Nocardioides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>0.01</td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>142577</td>\n",
       "      <td>Prauserella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>842</th>\n",
       "      <td>0.01</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>1868332</td>\n",
       "      <td>Neomicrococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950</th>\n",
       "      <td>0.01</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>57495</td>\n",
       "      <td>Dermacoccus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888</th>\n",
       "      <td>0.01</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>518733</td>\n",
       "      <td>Microterricola</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>0.01</td>\n",
       "      <td>41</td>\n",
       "      <td>4</td>\n",
       "      <td>G</td>\n",
       "      <td>1742993</td>\n",
       "      <td>Pseudarthrobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>756</th>\n",
       "      <td>0.01</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>1827</td>\n",
       "      <td>Rhodococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>0.01</td>\n",
       "      <td>106</td>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>1279</td>\n",
       "      <td>Staphylococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>0.01</td>\n",
       "      <td>36</td>\n",
       "      <td>33</td>\n",
       "      <td>G</td>\n",
       "      <td>1375</td>\n",
       "      <td>Aerococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>0.01</td>\n",
       "      <td>50</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "      <td>1578</td>\n",
       "      <td>Lactobacillus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1571</th>\n",
       "      <td>0.01</td>\n",
       "      <td>39</td>\n",
       "      <td>14</td>\n",
       "      <td>G</td>\n",
       "      <td>374</td>\n",
       "      <td>Bradyrhizobium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Percent  CumulCount  Count TaxLevel    Taxid                 Name\n",
       "31      53.03      379055      0        G     9605                 Homo\n",
       "451     16.64      118902  98625        G     1763        Mycobacterium\n",
       "39      14.54      103925  63260        G     1301        Streptococcus\n",
       "212      9.83       70278  37791        G     1378              Gemella\n",
       "807      1.57       11236    184        G    32207               Rothia\n",
       "373      0.33        2375   1513        G    29465          Veillonella\n",
       "813      0.12         848    359        G     1269          Micrococcus\n",
       "683      0.09         644    104        G     1716      Corynebacterium\n",
       "218      0.07         517     28        G     1386             Bacillus\n",
       "817      0.05         328    121        G    57493              Kocuria\n",
       "630      0.05         377     38        G  1866885    Mycolicibacterium\n",
       "827      0.04         266     14        G     1663         Arthrobacter\n",
       "1037     0.04         266      4        G  1912216        Cutibacterium\n",
       "1095     0.03         200     22        G     1883         Streptomyces\n",
       "869      0.02         120    102        G    33882       Microbacterium\n",
       "885      0.02         118      0        G   529881  Candidatus Aquiluna\n",
       "942      0.02         110    110        G   157920   Cellulosimicrobium\n",
       "996      0.01          38      0        G     1696       Brevibacterium\n",
       "1185     0.01          85      7        G     1654          Actinomyces\n",
       "1469     0.01          64      0        G       68           Lysobacter\n",
       "1481     0.01          36     10        G      286          Pseudomonas\n",
       "1065     0.01          47      7        G     1839         Nocardioides\n",
       "1148     0.01          90      0        G   142577          Prauserella\n",
       "842      0.01          53      0        G  1868332       Neomicrococcus\n",
       "950      0.01          95      0        G    57495          Dermacoccus\n",
       "888      0.01          44      0        G   518733       Microterricola\n",
       "844      0.01          41      4        G  1742993    Pseudarthrobacter\n",
       "756      0.01          81      8        G     1827          Rhodococcus\n",
       "299      0.01         106     16        G     1279       Staphylococcus\n",
       "188      0.01          36     33        G     1375           Aerococcus\n",
       "164      0.01          50      8        G     1578        Lactobacillus\n",
       "1571     0.01          39     14        G      374       Bradyrhizobium"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kraken.query(\"Percent > 0 & TaxLevel in ['G']\").sort_values('Percent', ascending=False)#.Name.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13235b88-9f63-4840-b1d5-f05166fbdb96",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "021d41b2-b65e-4a15-a222-7e67c23be037",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_found_DS_samples = list(set(DS_samples) - set(df_trust_WGS_metadata['pid']))\n",
    "len(not_found_DS_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c2fe1b19-90d5-424e-8b92-9653d73a399d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['T0454', 'T0438', 'T0281', 'T0107Cntrl', 'T0476', 'K064773', 'T0348', 'H37']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_found_DS_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "5fd93be6-e516-4bd9-92da-fe57c7d2204c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def combine_TRUST_patient_samples(df_trust_patient_data, WGS_metadata):\n",
    "\n",
    "    print(f\"{len(df_trust_patient_data.patient_num.unique())} patients in phenotypes dataframe\")\n",
    "    \n",
    "    # each sample ID is in the format like 260-04, S0299-01, or S0271-01A. So use re to remove all alpha characters and keep only numbers.\n",
    "    # first number is the patient number, second is the sample number\n",
    "    # merge the patient number extracted from each sample with the patient numbers in df_trust_patient_data\n",
    "    for i, row in WGS_metadata.iterrows():\n",
    "        try:\n",
    "            patient_num = int(re.findall(r'\\d+', row[\"Original_ID\"].split(\"-\")[0])[0])\n",
    "        # weird names like T-7, T-5, and P-T- 1\n",
    "        except:\n",
    "            patient_num = row[\"Original_ID\"]\n",
    "\n",
    "        try:\n",
    "            sample_week = int(re.findall(r'\\d+', row[\"Original_ID\"].split(\"-\")[1])[0])\n",
    "        # weird names like T-7, T-5, and P-T- 1\n",
    "        except:\n",
    "            sample_week = row['Original_ID']\n",
    "            \n",
    "        WGS_metadata.loc[i, ['patient_num', 'sample_week']] = [patient_num, sample_week]\n",
    "            \n",
    "    # WGS_metadata[\"patient_num\"] = [int(re.findall(r'\\d+', val.split(\"-\")[0])[0]) for val in WGS_metadata[\"Original_ID\"].values]\n",
    "    # WGS_metadata[\"sample_week\"] = [int(re.findall(r'\\d+', val.split(\"-\")[1])[0]) for val in WGS_metadata[\"Original_ID\"].values]\n",
    "\n",
    "    # keep only samples for the patients in df_trust_patient_data\n",
    "    WGS_metadata = WGS_metadata.merge(df_trust_patient_data, on=\"patient_num\", how=\"inner\")\n",
    "    print(f\"Found {WGS_metadata.patient_num.nunique()}/{df_trust_patient_data.patient_num.nunique()} patients in the WGS metadata files\")\n",
    "    \n",
    "    # rename some columns that have spaces in them\n",
    "    WGS_metadata = WGS_metadata.rename(columns={\"Cov Any Mean\": \"Cov_Any_Mean\",\n",
    "                                                \"Cov Unam Perc\": \"Cov_Unam_Perc\",\n",
    "                                                \"Perc. Reads Mapped\": \"Perc_Reads_Mapped\",\n",
    "                                                \"phylogenetic classification (Coll et al., 2014)\": \"Coll2014_Annotated\"\n",
    "                                       })\n",
    "\n",
    "    return WGS_metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "68465129-790e-4d45-bccc-764b767dc560",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Original_ID</th>\n",
       "      <th>shipment</th>\n",
       "      <th>status</th>\n",
       "      <th>comment</th>\n",
       "      <th>SampleLib</th>\n",
       "      <th>phylogenetic classification (Coll et al., 2014)</th>\n",
       "      <th>quality</th>\n",
       "      <th>Cov Any Mean</th>\n",
       "      <th>Cov Unam Perc</th>\n",
       "      <th>Perc. Reads Mapped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFS-179</td>\n",
       "      <td>260-04</td>\n",
       "      <td>1st</td>\n",
       "      <td>failed</td>\n",
       "      <td>heavily contaminated</td>\n",
       "      <td>MFS-179_lib48359</td>\n",
       "      <td>2.2.1.1 Beijing</td>\n",
       "      <td>4</td>\n",
       "      <td>6.65</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFS-78</td>\n",
       "      <td>157-01</td>\n",
       "      <td>1st</td>\n",
       "      <td>failed</td>\n",
       "      <td>heavily contaminated</td>\n",
       "      <td>MFS-78_lib47065</td>\n",
       "      <td>4.6.2.1 Euro-American Ancestral 2</td>\n",
       "      <td>3</td>\n",
       "      <td>147.41</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.4469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFS-183</td>\n",
       "      <td>264-07</td>\n",
       "      <td>1st</td>\n",
       "      <td>failed</td>\n",
       "      <td>heavily contaminated</td>\n",
       "      <td>MFS-183_lib48363</td>\n",
       "      <td>2.2.1 Beijing</td>\n",
       "      <td>1</td>\n",
       "      <td>186.20</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.6105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFS-141</td>\n",
       "      <td>231-05</td>\n",
       "      <td>1st</td>\n",
       "      <td>finished</td>\n",
       "      <td>-</td>\n",
       "      <td>MFS-141_lib47320</td>\n",
       "      <td>4.3.2.1 LAM</td>\n",
       "      <td>1</td>\n",
       "      <td>268.19</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.9833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFS-188</td>\n",
       "      <td>268-09</td>\n",
       "      <td>1st</td>\n",
       "      <td>finished</td>\n",
       "      <td>contaminated</td>\n",
       "      <td>MFS-188_lib48368</td>\n",
       "      <td>2.2.1 Beijing Asian/Africa 2</td>\n",
       "      <td>1</td>\n",
       "      <td>282.35</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.8306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>MFS-284</td>\n",
       "      <td>S0295-05</td>\n",
       "      <td>2nd</td>\n",
       "      <td>finished</td>\n",
       "      <td>-</td>\n",
       "      <td>MFS-284_lib52225</td>\n",
       "      <td>2.2.1.1 Beijing Pacific RD150</td>\n",
       "      <td>1</td>\n",
       "      <td>234.63</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>MFS-285</td>\n",
       "      <td>S0297-01</td>\n",
       "      <td>2nd</td>\n",
       "      <td>finished</td>\n",
       "      <td>-</td>\n",
       "      <td>MFS-285_lib52226</td>\n",
       "      <td>4.3.4.1 LAM</td>\n",
       "      <td>1</td>\n",
       "      <td>233.53</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>MFS-302</td>\n",
       "      <td>S0286-08</td>\n",
       "      <td>2nd</td>\n",
       "      <td>finished</td>\n",
       "      <td>-</td>\n",
       "      <td>MFS-302_lib52243</td>\n",
       "      <td>2.2.1.1 Beijing Pacific RD150</td>\n",
       "      <td>1</td>\n",
       "      <td>215.07</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>MFS-259</td>\n",
       "      <td>S0299-01</td>\n",
       "      <td>2nd</td>\n",
       "      <td>finished</td>\n",
       "      <td>-</td>\n",
       "      <td>MFS-259_lib51729</td>\n",
       "      <td>2.2.1 Beijing Asian/Africa 2</td>\n",
       "      <td>1</td>\n",
       "      <td>184.20</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>MFS-304</td>\n",
       "      <td>S0287-06</td>\n",
       "      <td>2nd</td>\n",
       "      <td>finished</td>\n",
       "      <td>-</td>\n",
       "      <td>MFS-304_lib52345</td>\n",
       "      <td>2.2.1.1 Beijing Pacific RD150</td>\n",
       "      <td>1</td>\n",
       "      <td>153.13</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.4615</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    SampleID Original_ID shipment    status               comment  \\\n",
       "0    MFS-179      260-04      1st    failed  heavily contaminated   \n",
       "1     MFS-78      157-01      1st    failed  heavily contaminated   \n",
       "2    MFS-183      264-07      1st    failed  heavily contaminated   \n",
       "3    MFS-141      231-05      1st  finished                     -   \n",
       "4    MFS-188      268-09      1st  finished          contaminated   \n",
       "..       ...         ...      ...       ...                   ...   \n",
       "262  MFS-284    S0295-05      2nd  finished                     -   \n",
       "263  MFS-285    S0297-01      2nd  finished                     -   \n",
       "264  MFS-302    S0286-08      2nd  finished                     -   \n",
       "265  MFS-259    S0299-01      2nd  finished                     -   \n",
       "266  MFS-304    S0287-06      2nd  finished                     -   \n",
       "\n",
       "            SampleLib phylogenetic classification (Coll et al., 2014)  \\\n",
       "0    MFS-179_lib48359                                 2.2.1.1 Beijing   \n",
       "1     MFS-78_lib47065               4.6.2.1 Euro-American Ancestral 2   \n",
       "2    MFS-183_lib48363                                   2.2.1 Beijing   \n",
       "3    MFS-141_lib47320                                     4.3.2.1 LAM   \n",
       "4    MFS-188_lib48368                    2.2.1 Beijing Asian/Africa 2   \n",
       "..                ...                                             ...   \n",
       "262  MFS-284_lib52225                   2.2.1.1 Beijing Pacific RD150   \n",
       "263  MFS-285_lib52226                                     4.3.4.1 LAM   \n",
       "264  MFS-302_lib52243                   2.2.1.1 Beijing Pacific RD150   \n",
       "265  MFS-259_lib51729                    2.2.1 Beijing Asian/Africa 2   \n",
       "266  MFS-304_lib52345                   2.2.1.1 Beijing Pacific RD150   \n",
       "\n",
       "     quality  Cov Any Mean  Cov Unam Perc  Perc. Reads Mapped  \n",
       "0          4          6.65           0.00              0.0916  \n",
       "1          3        147.41           0.70              0.4469  \n",
       "2          1        186.20           0.97              0.6105  \n",
       "3          1        268.19           1.00              0.9833  \n",
       "4          1        282.35           0.99              0.8306  \n",
       "..       ...           ...            ...                 ...  \n",
       "262        1        234.63           0.99              0.9873  \n",
       "263        1        233.53           0.99              0.9900  \n",
       "264        1        215.07           0.99              0.9916  \n",
       "265        1        184.20           0.99              0.9866  \n",
       "266        1        153.13           0.99              0.4615  \n",
       "\n",
       "[267 rows x 11 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "1c6549f8-b6ff-4463-b8de-4c955c297bf3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 WGS metadata Excel files\n",
      "458 patients in phenotypes dataframe\n",
      "Found 0/458 patients in the WGS metadata files\n"
     ]
    }
   ],
   "source": [
    "patient_data_fName = \"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/clinical_data/TRUST_DATA_2025-05-12_1129.cleaned.wide.csv\"\n",
    "WGS_report_directory = \"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS_metadata_reports\"\n",
    "WGS_data_fName = \"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/Illumina_culture_WGS_summary.csv\"\n",
    "\n",
    "df_trust_patient_data = pd.read_csv(patient_data_fName, low_memory=False)\n",
    "\n",
    "# get the patient number to match WGS IDs and pids\n",
    "df_trust_patient_data[\"patient_num\"] = [int(patient_id.replace(\"T0\", \"\")) for patient_id in df_trust_patient_data[\"pid\"].values]\n",
    "\n",
    "# get all Excel files from this directory\n",
    "trust_report_fNames = glob.glob(f\"{WGS_report_directory}/*.xlsx\")\n",
    "print(f\"{len(trust_report_fNames)} WGS metadata Excel files\")\n",
    "\n",
    "df_trust_WGS_metadata = []\n",
    "\n",
    "# sort chronologically because below, we preferentially keep the later one\n",
    "for fName in np.sort(trust_report_fNames)[::-1]:\n",
    "\n",
    "    # read in single Excel file\n",
    "    df = pd.read_excel(fName, sheet_name=None)\n",
    "\n",
    "    # remove spaces from column name to make querying easier. Also there could be NaN rows if there are additional empty rows in the Excel sheet, so drop them\n",
    "    df = df['summary'].rename(columns={'original ID': 'Original_ID'}).dropna(axis=0, how='all')\n",
    "\n",
    "    # append to list for concatenation\n",
    "    df_trust_WGS_metadata.append(df)\n",
    "\n",
    "# the Excel files above are running totals, so the most recent file has data that is also in the older files. So drop duplicates, keeping the most recent (last) one\n",
    "df_trust_WGS_metadata = pd.concat(df_trust_WGS_metadata).drop_duplicates('SampleID', keep='last')#.query(\"status!='failed'\")\n",
    "\n",
    "# combine patient and sample IDs (pid = patient, Original_ID and SampleID = WGS)\n",
    "df_trust_combined = combine_TRUST_patient_samples(df_trust_patient_data, df_trust_WGS_metadata.reset_index(drop=True))\n",
    "\n",
    "# combine with additional WGS QC data. Drop samples that filed QC (no VCF, so they have nothing in the Lineage column)\n",
    "df_geno = pd.read_csv(WGS_data_fName).dropna(subset='Lineage')\n",
    "\n",
    "# df_trust_combined = df_trust_combined.merge(df_geno[['SampleID', 'F2', 'Coll2014', 'Freschi2020', 'Lineage']], on='SampleID', how='left').reset_index(drop=True)\n",
    "\n",
    "# # Remove columns that are NaN everywhere to clean the dataframe\n",
    "# df_trust_combined = df_trust_combined.dropna(how='all', axis=1)\n",
    "\n",
    "# # get the sample week\n",
    "# df_trust_combined['Sampling_Week'] = df_trust_combined['Original_ID'].str.split('-').str[1]\n",
    "\n",
    "# # replace month 5 with 20 for weeks\n",
    "# df_trust_combined['Sampling_Week'] = df_trust_combined['Sampling_Week'].replace('01A', '01').replace('m5', '20').replace('M5', '20')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "58b2c962-f2d5-4ecf-8e99-4c4c95784e0c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS_metadata_reports/2023-01-24_report_Farhat.v09.xlsx',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS_metadata_reports/2023-09-01_report_Farhat.v09.xlsx',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS_metadata_reports/2024-05-27_report.v10.xlsx',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS_metadata_reports/2024-06-25_report.v10.xlsx',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS_metadata_reports/2024-11-13_report.v10.xlsx',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS_metadata_reports/2025-07-25_report.v10.xlsx']"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trust_report_fNames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "92de7ee6-fb1f-461b-a826-7e2db0080c62",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# read in single Excel file\n",
    "# df = pd.read_excel('/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS_metadata_reports/2025-07-25_report.v10.xlsx', sheet_name=None)\n",
    "df = pd.read_excel('/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Illumina/250730_IlluminaWGS_lastbatch/2025-07-25_report.v10.xlsx', sheet_name=None)\n",
    "\n",
    "# remove spaces from column name to make querying easier. Also there could be NaN rows if there are additional empty rows in the Excel sheet, so drop them\n",
    "df = df['summary'].rename(columns={'original ID': 'Original_ID'}).dropna(axis=0, how='all')\n",
    "\n",
    "df['Original_ID'] = df['Original_ID'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "98983070-daa6-4aab-81a7-bd10ad05bba3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>SampleLib</th>\n",
       "      <th>Original_ID</th>\n",
       "      <th>Pacbio</th>\n",
       "      <th>Comment</th>\n",
       "      <th>quality</th>\n",
       "      <th>Cov Any Mean</th>\n",
       "      <th>Cov Unam Perc</th>\n",
       "      <th>Perc. Reads Mapped</th>\n",
       "      <th>Excluded</th>\n",
       "      <th>repeat_pcrfree_1</th>\n",
       "      <th>repeat_pcrfree_2</th>\n",
       "      <th>transfered_pcrfree</th>\n",
       "      <th>transfered_pacbio</th>\n",
       "      <th>Repeat_1_factor</th>\n",
       "      <th>repat_1_loading</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MFS-918</td>\n",
       "      <td>MFS-918_lib84065</td>\n",
       "      <td>TS373</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>283.69</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SampleID         SampleLib Original_ID  Pacbio Comment quality  \\\n",
       "18  MFS-918  MFS-918_lib84065       TS373     NaN     NaN       1   \n",
       "\n",
       "   Cov Any Mean Cov Unam Perc Perc. Reads Mapped Excluded repeat_pcrfree_1  \\\n",
       "18       283.69             1             0.9846        0              NaN   \n",
       "\n",
       "   repeat_pcrfree_2 transfered_pcrfree transfered_pacbio Repeat_1_factor  \\\n",
       "18              NaN                  0                 0             NaN   \n",
       "\n",
       "   repat_1_loading  \n",
       "18             NaN  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.query(\"SampleID=='MFS-918'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "e65bb768-00b4-46c7-829e-a7f0c0e62163",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>SampleLib</th>\n",
       "      <th>Original_ID</th>\n",
       "      <th>Pacbio</th>\n",
       "      <th>Comment</th>\n",
       "      <th>quality</th>\n",
       "      <th>Cov Any Mean</th>\n",
       "      <th>Cov Unam Perc</th>\n",
       "      <th>Perc. Reads Mapped</th>\n",
       "      <th>Excluded</th>\n",
       "      <th>repeat_pcrfree_1</th>\n",
       "      <th>repeat_pcrfree_2</th>\n",
       "      <th>transfered_pcrfree</th>\n",
       "      <th>transfered_pacbio</th>\n",
       "      <th>Repeat_1_factor</th>\n",
       "      <th>repat_1_loading</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MFS-918</td>\n",
       "      <td>MFS-918_lib84065</td>\n",
       "      <td>TS373</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>283.69</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MFS-919</td>\n",
       "      <td>MFS-919_lib84066</td>\n",
       "      <td>TS374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>273.82</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9812</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MFS-911</td>\n",
       "      <td>MFS-911_lib84058</td>\n",
       "      <td>TS95</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>263.04</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9821</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MFS-923</td>\n",
       "      <td>MFS-923_lib84070</td>\n",
       "      <td>TS586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>313.35</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9851</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.106431</td>\n",
       "      <td>-0.319292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MFS-921</td>\n",
       "      <td>MFS-921_lib84068</td>\n",
       "      <td>TS465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>291.56</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9852</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.039649</td>\n",
       "      <td>-0.118946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MFS-912</td>\n",
       "      <td>MFS-912_lib84059</td>\n",
       "      <td>TS128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>280.54</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9834</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.001925</td>\n",
       "      <td>-0.005775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MFS-914</td>\n",
       "      <td>MFS-914_lib84061</td>\n",
       "      <td>TS168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>269.2</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9858</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.040119</td>\n",
       "      <td>0.120357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>MFS-913</td>\n",
       "      <td>MFS-913_lib84060</td>\n",
       "      <td>TS147</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>329.17</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9844</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.149376</td>\n",
       "      <td>-0.448127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>MFS-910</td>\n",
       "      <td>MFS-910_lib84057</td>\n",
       "      <td>TS12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>312.2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9842</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.103139</td>\n",
       "      <td>-0.309417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>MFS-915</td>\n",
       "      <td>MFS-915_lib84062</td>\n",
       "      <td>TS183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>314.4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9841</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.109415</td>\n",
       "      <td>-0.328244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>MFS-922</td>\n",
       "      <td>MFS-922_lib84069</td>\n",
       "      <td>TS522</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>307.64</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9843</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.089845</td>\n",
       "      <td>-0.269536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>MFS-916</td>\n",
       "      <td>MFS-916_lib84063</td>\n",
       "      <td>TS184</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>295.09</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.051137</td>\n",
       "      <td>-0.153411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>MFS-917</td>\n",
       "      <td>MFS-917_lib84064</td>\n",
       "      <td>TS310</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>301.21</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9802</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.070416</td>\n",
       "      <td>-0.211248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>MFS-920</td>\n",
       "      <td>MFS-920_lib84067</td>\n",
       "      <td>TS405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>294.23</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9829</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.048364</td>\n",
       "      <td>-0.145091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SampleID         SampleLib Original_ID  Pacbio Comment quality  \\\n",
       "18  MFS-918  MFS-918_lib84065       TS373     NaN     NaN       1   \n",
       "22  MFS-919  MFS-919_lib84066       TS374     NaN     NaN       1   \n",
       "25  MFS-911  MFS-911_lib84058        TS95     NaN     NaN       1   \n",
       "29  MFS-923  MFS-923_lib84070       TS586     NaN     NaN       1   \n",
       "33  MFS-921  MFS-921_lib84068       TS465     NaN     NaN       1   \n",
       "37  MFS-912  MFS-912_lib84059       TS128     NaN     NaN       1   \n",
       "38  MFS-914  MFS-914_lib84061       TS168     NaN     NaN       1   \n",
       "40  MFS-913  MFS-913_lib84060       TS147     NaN     NaN       1   \n",
       "42  MFS-910  MFS-910_lib84057        TS12     NaN     NaN       1   \n",
       "43  MFS-915  MFS-915_lib84062       TS183     NaN     NaN       1   \n",
       "44  MFS-922  MFS-922_lib84069       TS522     NaN     NaN       1   \n",
       "46  MFS-916  MFS-916_lib84063       TS184     NaN     NaN       1   \n",
       "47  MFS-917  MFS-917_lib84064       TS310     NaN     NaN       1   \n",
       "51  MFS-920  MFS-920_lib84067       TS405     NaN     NaN       1   \n",
       "\n",
       "   Cov Any Mean Cov Unam Perc Perc. Reads Mapped Excluded repeat_pcrfree_1  \\\n",
       "18       283.69             1             0.9846        0              NaN   \n",
       "22       273.82          0.99             0.9812        0              NaN   \n",
       "25       263.04          0.99             0.9821        0              NaN   \n",
       "29       313.35          0.99             0.9851        0                1   \n",
       "33       291.56          0.99             0.9852        0                1   \n",
       "37       280.54          0.99             0.9834        0                1   \n",
       "38        269.2          0.99             0.9858        0                1   \n",
       "40       329.17          0.99             0.9844        0                1   \n",
       "42        312.2             1             0.9842        0                1   \n",
       "43        314.4             1             0.9841        0                1   \n",
       "44       307.64             1             0.9843        0                1   \n",
       "46       295.09          0.99             0.9846        0                1   \n",
       "47       301.21          0.99             0.9802        0                1   \n",
       "51       294.23          0.99             0.9829        0                1   \n",
       "\n",
       "   repeat_pcrfree_2 transfered_pcrfree transfered_pacbio Repeat_1_factor  \\\n",
       "18              NaN                  0                 0             NaN   \n",
       "22              NaN                  0                 0             NaN   \n",
       "25              NaN                  0                 0             NaN   \n",
       "29              NaN                  0                 0       -0.106431   \n",
       "33              NaN                  0                 0       -0.039649   \n",
       "37              NaN                  0                 0       -0.001925   \n",
       "38              NaN                  0                 0        0.040119   \n",
       "40              NaN                  0                 0       -0.149376   \n",
       "42              NaN                  0                 0       -0.103139   \n",
       "43              NaN                  0                 0       -0.109415   \n",
       "44              NaN                  0                 0       -0.089845   \n",
       "46              NaN                  0                 0       -0.051137   \n",
       "47              NaN                  0                 0       -0.070416   \n",
       "51              NaN                  0                 0       -0.048364   \n",
       "\n",
       "   repat_1_loading  \n",
       "18             NaN  \n",
       "22             NaN  \n",
       "25             NaN  \n",
       "29       -0.319292  \n",
       "33       -0.118946  \n",
       "37       -0.005775  \n",
       "38        0.120357  \n",
       "40       -0.448127  \n",
       "42       -0.309417  \n",
       "43       -0.328244  \n",
       "44       -0.269536  \n",
       "46       -0.153411  \n",
       "47       -0.211248  \n",
       "51       -0.145091  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.query(\"Original_ID.str.contains('TS')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "2defef39-1732-44be-b76a-99aa707b2438",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>SampleLib</th>\n",
       "      <th>Original_ID</th>\n",
       "      <th>Pacbio</th>\n",
       "      <th>Comment</th>\n",
       "      <th>quality</th>\n",
       "      <th>Cov Any Mean</th>\n",
       "      <th>Cov Unam Perc</th>\n",
       "      <th>Perc. Reads Mapped</th>\n",
       "      <th>Excluded</th>\n",
       "      <th>repeat_pcrfree_1</th>\n",
       "      <th>repeat_pcrfree_2</th>\n",
       "      <th>transfered_pcrfree</th>\n",
       "      <th>transfered_pacbio</th>\n",
       "      <th>Repeat_1_factor</th>\n",
       "      <th>repat_1_loading</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MFS-918</td>\n",
       "      <td>MFS-918_lib84065</td>\n",
       "      <td>TS373</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>283.69</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MFS-919</td>\n",
       "      <td>MFS-919_lib84066</td>\n",
       "      <td>TS374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>273.82</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9812</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MFS-911</td>\n",
       "      <td>MFS-911_lib84058</td>\n",
       "      <td>TS95</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>263.04</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9821</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MFS-923</td>\n",
       "      <td>MFS-923_lib84070</td>\n",
       "      <td>TS586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>313.35</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9851</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.106431</td>\n",
       "      <td>-0.319292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MFS-921</td>\n",
       "      <td>MFS-921_lib84068</td>\n",
       "      <td>TS465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>291.56</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9852</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.039649</td>\n",
       "      <td>-0.118946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MFS-912</td>\n",
       "      <td>MFS-912_lib84059</td>\n",
       "      <td>TS128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>280.54</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9834</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.001925</td>\n",
       "      <td>-0.005775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MFS-914</td>\n",
       "      <td>MFS-914_lib84061</td>\n",
       "      <td>TS168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>269.2</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9858</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.040119</td>\n",
       "      <td>0.120357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>MFS-913</td>\n",
       "      <td>MFS-913_lib84060</td>\n",
       "      <td>TS147</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>329.17</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9844</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.149376</td>\n",
       "      <td>-0.448127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>MFS-910</td>\n",
       "      <td>MFS-910_lib84057</td>\n",
       "      <td>TS12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>312.2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9842</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.103139</td>\n",
       "      <td>-0.309417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>MFS-915</td>\n",
       "      <td>MFS-915_lib84062</td>\n",
       "      <td>TS183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>314.4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9841</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.109415</td>\n",
       "      <td>-0.328244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>MFS-922</td>\n",
       "      <td>MFS-922_lib84069</td>\n",
       "      <td>TS522</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>307.64</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9843</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.089845</td>\n",
       "      <td>-0.269536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>MFS-916</td>\n",
       "      <td>MFS-916_lib84063</td>\n",
       "      <td>TS184</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>295.09</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9846</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.051137</td>\n",
       "      <td>-0.153411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>MFS-917</td>\n",
       "      <td>MFS-917_lib84064</td>\n",
       "      <td>TS310</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>301.21</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9802</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.070416</td>\n",
       "      <td>-0.211248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>MFS-920</td>\n",
       "      <td>MFS-920_lib84067</td>\n",
       "      <td>TS405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>294.23</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9829</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.048364</td>\n",
       "      <td>-0.145091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SampleID         SampleLib Original_ID  Pacbio Comment quality  \\\n",
       "18  MFS-918  MFS-918_lib84065       TS373     NaN     NaN       1   \n",
       "22  MFS-919  MFS-919_lib84066       TS374     NaN     NaN       1   \n",
       "25  MFS-911  MFS-911_lib84058        TS95     NaN     NaN       1   \n",
       "29  MFS-923  MFS-923_lib84070       TS586     NaN     NaN       1   \n",
       "33  MFS-921  MFS-921_lib84068       TS465     NaN     NaN       1   \n",
       "37  MFS-912  MFS-912_lib84059       TS128     NaN     NaN       1   \n",
       "38  MFS-914  MFS-914_lib84061       TS168     NaN     NaN       1   \n",
       "40  MFS-913  MFS-913_lib84060       TS147     NaN     NaN       1   \n",
       "42  MFS-910  MFS-910_lib84057        TS12     NaN     NaN       1   \n",
       "43  MFS-915  MFS-915_lib84062       TS183     NaN     NaN       1   \n",
       "44  MFS-922  MFS-922_lib84069       TS522     NaN     NaN       1   \n",
       "46  MFS-916  MFS-916_lib84063       TS184     NaN     NaN       1   \n",
       "47  MFS-917  MFS-917_lib84064       TS310     NaN     NaN       1   \n",
       "51  MFS-920  MFS-920_lib84067       TS405     NaN     NaN       1   \n",
       "\n",
       "   Cov Any Mean Cov Unam Perc Perc. Reads Mapped Excluded repeat_pcrfree_1  \\\n",
       "18       283.69             1             0.9846        0              NaN   \n",
       "22       273.82          0.99             0.9812        0              NaN   \n",
       "25       263.04          0.99             0.9821        0              NaN   \n",
       "29       313.35          0.99             0.9851        0                1   \n",
       "33       291.56          0.99             0.9852        0                1   \n",
       "37       280.54          0.99             0.9834        0                1   \n",
       "38        269.2          0.99             0.9858        0                1   \n",
       "40       329.17          0.99             0.9844        0                1   \n",
       "42        312.2             1             0.9842        0                1   \n",
       "43        314.4             1             0.9841        0                1   \n",
       "44       307.64             1             0.9843        0                1   \n",
       "46       295.09          0.99             0.9846        0                1   \n",
       "47       301.21          0.99             0.9802        0                1   \n",
       "51       294.23          0.99             0.9829        0                1   \n",
       "\n",
       "   repeat_pcrfree_2 transfered_pcrfree transfered_pacbio Repeat_1_factor  \\\n",
       "18              NaN                  0                 0             NaN   \n",
       "22              NaN                  0                 0             NaN   \n",
       "25              NaN                  0                 0             NaN   \n",
       "29              NaN                  0                 0       -0.106431   \n",
       "33              NaN                  0                 0       -0.039649   \n",
       "37              NaN                  0                 0       -0.001925   \n",
       "38              NaN                  0                 0        0.040119   \n",
       "40              NaN                  0                 0       -0.149376   \n",
       "42              NaN                  0                 0       -0.103139   \n",
       "43              NaN                  0                 0       -0.109415   \n",
       "44              NaN                  0                 0       -0.089845   \n",
       "46              NaN                  0                 0       -0.051137   \n",
       "47              NaN                  0                 0       -0.070416   \n",
       "51              NaN                  0                 0       -0.048364   \n",
       "\n",
       "   repat_1_loading  \n",
       "18             NaN  \n",
       "22             NaN  \n",
       "25             NaN  \n",
       "29       -0.319292  \n",
       "33       -0.118946  \n",
       "37       -0.005775  \n",
       "38        0.120357  \n",
       "40       -0.448127  \n",
       "42       -0.309417  \n",
       "43       -0.328244  \n",
       "44       -0.269536  \n",
       "46       -0.153411  \n",
       "47       -0.211248  \n",
       "51       -0.145091  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.query(\"Original_ID.str.contains('TS')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77106bc2-2bea-4a52-9303-54a67e66a4f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "84eba259-0f39-4cc0-aaab-29b03c4ea037",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_Anna = pd.read_csv(\"~/Mtb_Megapipe/isolates_for_Anna.tsv\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9e868ee4-be79-4a72-9012-38c5ebd1e8c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for i, row in df_Anna.iterrows():\n",
    "    \n",
    "    if os.path.isfile(f\"/n/data1/hms/dbmi/farhat/rollingDB/genomic_data/{row[0]}/pilon/{row[0]}_variants.vcf\"):\n",
    "        df_Anna.loc[i, 'Done'] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c829fb37-c76c-4f8e-a060-e218cf471afd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Done\n",
       "NaN    13985\n",
       "1.0     8470\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Anna.Done.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5597fcab-97a0-4690-9f4e-cd1c5c228624",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Done\n",
       "NaN    15367\n",
       "1.0     7088\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Anna.Done.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9dac1d06-8463-46a7-af03-05cc98cdf45c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "PZA_isolates = pd.read_csv(\"~/Mtb_Megapipe/PZA_isolates_5.tsv\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "988a7756-fba2-4152-a1f5-1e1ff20badfc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>SAMN08708749</td>\n",
       "      <td>SRR6831820</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                0           1  2\n",
       "240  SAMN08708749  SRR6831820  1"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PZA_isolates.loc[PZA_isolates[1]=='SRR6831820']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09c3d54d-314d-4fbd-a928-3d284a794b07",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2ef8d749-474c-4a26-81f6-0a8d4603b890",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_kraken = pd.read_csv(\"~/TRUST_lowAF_keep_kraken_proportion.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "525b1793-f74c-4e78-967e-ff9dd1a5162e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Total_Reads</th>\n",
       "      <th>Keep_Reads</th>\n",
       "      <th>Keep_Proportion</th>\n",
       "      <th>Kraken_Unclassified_Percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFS-605</td>\n",
       "      <td>6824968</td>\n",
       "      <td>5509747</td>\n",
       "      <td>0.807293</td>\n",
       "      <td>18.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFS-188</td>\n",
       "      <td>4621344</td>\n",
       "      <td>3825199</td>\n",
       "      <td>0.827724</td>\n",
       "      <td>17.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFS-366</td>\n",
       "      <td>6097976</td>\n",
       "      <td>5262188</td>\n",
       "      <td>0.862940</td>\n",
       "      <td>13.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFS-778</td>\n",
       "      <td>5448727</td>\n",
       "      <td>4736817</td>\n",
       "      <td>0.869344</td>\n",
       "      <td>12.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFS-409</td>\n",
       "      <td>4951776</td>\n",
       "      <td>4478437</td>\n",
       "      <td>0.904410</td>\n",
       "      <td>10.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MFS-138</td>\n",
       "      <td>5083364</td>\n",
       "      <td>4683315</td>\n",
       "      <td>0.921302</td>\n",
       "      <td>7.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MFS-397</td>\n",
       "      <td>5586986</td>\n",
       "      <td>5283277</td>\n",
       "      <td>0.945640</td>\n",
       "      <td>4.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MFS-158</td>\n",
       "      <td>4817209</td>\n",
       "      <td>4591883</td>\n",
       "      <td>0.953225</td>\n",
       "      <td>3.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MFS-776</td>\n",
       "      <td>4899988</td>\n",
       "      <td>4735179</td>\n",
       "      <td>0.966365</td>\n",
       "      <td>2.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MFS-157</td>\n",
       "      <td>3713859</td>\n",
       "      <td>3596537</td>\n",
       "      <td>0.968410</td>\n",
       "      <td>2.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MFS-167</td>\n",
       "      <td>5733657</td>\n",
       "      <td>5553905</td>\n",
       "      <td>0.968650</td>\n",
       "      <td>2.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MFS-751</td>\n",
       "      <td>7113507</td>\n",
       "      <td>6925060</td>\n",
       "      <td>0.973509</td>\n",
       "      <td>2.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MFS-468</td>\n",
       "      <td>5067172</td>\n",
       "      <td>4939753</td>\n",
       "      <td>0.974854</td>\n",
       "      <td>2.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MFS-163</td>\n",
       "      <td>6510028</td>\n",
       "      <td>6353789</td>\n",
       "      <td>0.976000</td>\n",
       "      <td>1.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MFS-165</td>\n",
       "      <td>4728281</td>\n",
       "      <td>4615882</td>\n",
       "      <td>0.976228</td>\n",
       "      <td>1.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MFS-394</td>\n",
       "      <td>5492751</td>\n",
       "      <td>5364301</td>\n",
       "      <td>0.976615</td>\n",
       "      <td>1.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MFS-186</td>\n",
       "      <td>4188202</td>\n",
       "      <td>4093168</td>\n",
       "      <td>0.977309</td>\n",
       "      <td>1.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MFS-82</td>\n",
       "      <td>5303432</td>\n",
       "      <td>5184281</td>\n",
       "      <td>0.977533</td>\n",
       "      <td>1.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MFS-122</td>\n",
       "      <td>6268386</td>\n",
       "      <td>6130917</td>\n",
       "      <td>0.978069</td>\n",
       "      <td>1.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MFS-312</td>\n",
       "      <td>4644594</td>\n",
       "      <td>4543316</td>\n",
       "      <td>0.978194</td>\n",
       "      <td>1.04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SampleID  Total_Reads  Keep_Reads  Keep_Proportion  \\\n",
       "0   MFS-605      6824968     5509747         0.807293   \n",
       "1   MFS-188      4621344     3825199         0.827724   \n",
       "2   MFS-366      6097976     5262188         0.862940   \n",
       "3   MFS-778      5448727     4736817         0.869344   \n",
       "4   MFS-409      4951776     4478437         0.904410   \n",
       "5   MFS-138      5083364     4683315         0.921302   \n",
       "6   MFS-397      5586986     5283277         0.945640   \n",
       "7   MFS-158      4817209     4591883         0.953225   \n",
       "8   MFS-776      4899988     4735179         0.966365   \n",
       "9   MFS-157      3713859     3596537         0.968410   \n",
       "10  MFS-167      5733657     5553905         0.968650   \n",
       "11  MFS-751      7113507     6925060         0.973509   \n",
       "12  MFS-468      5067172     4939753         0.974854   \n",
       "13  MFS-163      6510028     6353789         0.976000   \n",
       "14  MFS-165      4728281     4615882         0.976228   \n",
       "15  MFS-394      5492751     5364301         0.976615   \n",
       "16  MFS-186      4188202     4093168         0.977309   \n",
       "17   MFS-82      5303432     5184281         0.977533   \n",
       "18  MFS-122      6268386     6130917         0.978069   \n",
       "19  MFS-312      4644594     4543316         0.978194   \n",
       "\n",
       "    Kraken_Unclassified_Percent  \n",
       "0                         18.03  \n",
       "1                         17.48  \n",
       "2                         13.22  \n",
       "3                         12.82  \n",
       "4                         10.60  \n",
       "5                          7.09  \n",
       "6                          4.99  \n",
       "7                          3.87  \n",
       "8                          2.69  \n",
       "9                          2.27  \n",
       "10                         2.24  \n",
       "11                         2.27  \n",
       "12                         2.02  \n",
       "13                         1.50  \n",
       "14                         1.52  \n",
       "15                         1.62  \n",
       "16                         1.28  \n",
       "17                         1.73  \n",
       "18                         1.59  \n",
       "19                         1.04  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kraken.sort_values(\"Keep_Proportion\").merge(df_WGS[['SampleID', 'Kraken_Unclassified_Percent']]).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84dbf197-48c5-4ec5-9b85-5546d3833098",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "10e8982c-aad3-42ec-a24a-e3ccbeeac557",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ROLLINGDB_ID</th>\n",
       "      <th>BIOSAMPLE_ACCESSION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>RUN</th>\n",
       "      <th>DB_OF_ORIGIN</th>\n",
       "      <th>ISOLATION_DATE</th>\n",
       "      <th>ISOLATION_COUNTRY</th>\n",
       "      <th>ISOLATION_REGION</th>\n",
       "      <th>SPUTUM/CULTURE</th>\n",
       "      <th>UNPAIRED/PAIRED</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>SPECIES</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Combined_Runs</th>\n",
       "      <th>Num_Runs</th>\n",
       "      <th>OtherNames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>C001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>C002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>C003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>C004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>C005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ROLLINGDB_ID BIOSAMPLE_ACCESSION SAMPLE  RUN DB_OF_ORIGIN ISOLATION_DATE  \\\n",
       "698         C001                 NaN    NaN  NaN         CASS            NaN   \n",
       "699         C002                 NaN    NaN  NaN         CASS            NaN   \n",
       "700         C003                 NaN    NaN  NaN         CASS            NaN   \n",
       "701         C004                 NaN    NaN  NaN         CASS            NaN   \n",
       "702         C005                 NaN    NaN  NaN         CASS            NaN   \n",
       "\n",
       "    ISOLATION_COUNTRY ISOLATION_REGION  SPUTUM/CULTURE  UNPAIRED/PAIRED  \\\n",
       "698               NaN              NaN               1                1   \n",
       "699               NaN              NaN               1                1   \n",
       "700               NaN              NaN               1                1   \n",
       "701               NaN              NaN               1                1   \n",
       "702               NaN              NaN               1                1   \n",
       "\n",
       "    Platform Model                     SPECIES  TaxID ReleaseDate LoadDate  \\\n",
       "698      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "699      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "700      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "701      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "702      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "\n",
       "    Combined_Runs  Num_Runs OtherNames  \n",
       "698           NaN       NaN        NaN  \n",
       "699           NaN       NaN        NaN  \n",
       "700           NaN       NaN        NaN  \n",
       "701           NaN       NaN        NaN  \n",
       "702           NaN       NaN        NaN  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_metadata.query(\"DB_OF_ORIGIN=='CASS'\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7dbfd12-48d4-4e86-bff3-dedafa8602af",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c8c07445-ff6e-406a-a870-4c43b61075f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "313"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_fastqs = glob.glob(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/C*\")\n",
    "len(cass_fastqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c8c0f3bc-bb4b-4a30-8443-14641c183af0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(313, 54)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_fastqs_isolates = [os.path.basename(fName) for fName in cass_fastqs]\n",
    "problem_cass_fastqs = [os.path.basename(fName) for fName in glob.glob(f\"/n/data1/hms/dbmi/farhat/rollingDB/problem_fastqs/C*\")]\n",
    "\n",
    "len(cass_fastqs_isolates), len(problem_cass_fastqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5b34b6ea-dc8e-4546-816f-3b58a088d88a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(54, 9)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_mapping_table.loc[cass_mapping_table['Sample Name'].isin(problem_cass_fastqs)][['BioSample', 'Run', 'Sample Name', 'geo_loc_name_country', 'geo_loc_name', 'Instrument', 'Platform', 'ReleaseDate', 'Collection_Date']].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8faee5ec-9fc2-4d1e-9550-be11dafffa91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(307, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Run</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>geo_loc_name_country</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Collection_Date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN13813810</td>\n",
       "      <td>SRR10868968</td>\n",
       "      <td>C429</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN13813725</td>\n",
       "      <td>SRR10868973</td>\n",
       "      <td>C094</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN13813804</td>\n",
       "      <td>SRR10868975</td>\n",
       "      <td>C412</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN13813803</td>\n",
       "      <td>SRR10868976</td>\n",
       "      <td>C408</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN13813802</td>\n",
       "      <td>SRR10868977</td>\n",
       "      <td>C407</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      BioSample          Run Sample Name geo_loc_name_country  geo_loc_name  \\\n",
       "0  SAMN13813810  SRR10868968        C429         South Africa  South Africa   \n",
       "1  SAMN13813725  SRR10868973        C094         South Africa  South Africa   \n",
       "2  SAMN13813804  SRR10868975        C412         South Africa  South Africa   \n",
       "3  SAMN13813803  SRR10868976        C408         South Africa  South Africa   \n",
       "4  SAMN13813802  SRR10868977        C407         South Africa  South Africa   \n",
       "\n",
       "            Instrument  Platform           ReleaseDate  Collection_Date  \n",
       "0  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "1  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "2  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "3  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "4  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_fastas = cass_mapping_table.loc[cass_mapping_table['Sample Name'].isin(cass_fastqs_isolates)][['BioSample', 'Run', 'Sample Name', 'geo_loc_name_country', 'geo_loc_name', 'Instrument', 'Platform', 'ReleaseDate', 'Collection_Date']].reset_index(drop=True)\n",
    "\n",
    "print(check_fastas.shape)\n",
    "\n",
    "check_fastas.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "80113d4a-849c-4be1-bf6a-0068efc4f2df",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_lines(file_path):\n",
    "    # Use subprocess to call wc -l to count the number of lines in a file\n",
    "    process = subprocess.Popen(['wc', '-l', file_path], stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    stdout, stderr = process.communicate()\n",
    "\n",
    "    if process.returncode != 0:\n",
    "        print(f\"Error counting lines in {file_path}: {stderr.decode().strip()}\")\n",
    "        return None\n",
    "\n",
    "    # Extract and return the line count from the output\n",
    "    line_count = int(stdout.decode().strip().split()[0])\n",
    "    return line_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5b1732b8-f898-410c-9810-f0da43b57562",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "50\n",
      "100\n",
      "150\n",
      "200\n",
      "250\n",
      "300\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(304, 3)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "good_files = []\n",
    "bad_files = []\n",
    "\n",
    "for i, row in check_fastas.iterrows():\n",
    "\n",
    "    internal_id = row['Sample Name']\n",
    "    run_id = row['Run']\n",
    "\n",
    "    # if not os.path.isfile(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_1.fastq\") or not os.path.isfile(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_2.fastq\"):\n",
    "\n",
    "    #     subprocess.run(f\"gunzip -c /n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_R1.fastq.gz > /n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_1.fastq\", shell=True)\n",
    "\n",
    "    #     subprocess.run(f\"gunzip -c /n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_R2.fastq.gz > /n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_2.fastq\", shell=True)\n",
    "\n",
    "    # if i % 25 == 0:\n",
    "    #     print(f\"Finished unzipping sample {i}\")\n",
    "        \n",
    "    F1_LC_internal = count_lines(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_1.fastq\")\n",
    "    F2_LC_internal = count_lines(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_2.fastq\")\n",
    "\n",
    "    if os.path.isfile(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_1.fastq\") and os.path.isfile(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_2.fastq\"):\n",
    "        \n",
    "        F1_LC_public = count_lines(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_1.fastq\")\n",
    "        F2_LC_public = count_lines(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_2.fastq\")\n",
    "\n",
    "        if F1_LC_internal == F2_LC_internal == F1_LC_public == F2_LC_public:\n",
    "            good_files.append(run_id)\n",
    "        else:\n",
    "            bad_files.append(run_id)\n",
    "\n",
    "    if i % 50 == 0:\n",
    "        print(i)\n",
    "\n",
    "len(good_files), len(bad_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4fdd0c29-c799-4a6f-87c5-22387912fe8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SRR10869030', 'SRR10869302', 'SRR10869310']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d141ed4e-7238-47bd-a1b0-f7ad53d4299b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1fbae39-def1-4603-ab0a-c3e1f34659e6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "240c9b32-0f77-4d0b-8fbd-8513596fb3f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_headers_FASTQ(fName):\n",
    "    subprocess.run(f\"awk 'NR % 2 == 0' {fName} > {fName.replace('_1', '_noID_1').replace('_2', '_noID_2')}\", shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e46566d3-427f-458a-8504-5909919a9712",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_files(file1, file2):\n",
    "    result = subprocess.run(['diff', file1, file2], stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    if result.returncode == 0:\n",
    "        # print(\"Files are the same.\")\n",
    "        return 0\n",
    "    else:\n",
    "        # print(\"Files are different.\")\n",
    "        return 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "89034051-2d63-4ddb-b92c-b78f4384902a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare_files('/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/C001/C001_noID_1.fastq', '/n/scratch/users/s/sak0914/smk_output/C001/SRR10869296/SRR10869296_noID_1.fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "03742e6f-f7f2-4baa-84b3-28870331cfaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare_files('/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/C429/C429_noID_1.fastq', '/n/scratch/users/s/sak0914/smk_output/C429/SRR10868968/SRR10868968_noID_1.fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2c41fdbc-ad1d-48d1-84df-e2badaf441e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>FQ1</th>\n",
       "      <th>FQ2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C429</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ID FQ1 FQ2\n",
       "0  C429   1   1"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_compare_FASTQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2734618b-0111-4adb-a89b-40ed03d37512",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df_compare_FASTQ = pd.DataFrame(columns=['ID', 'FQ1', 'FQ2'])\n",
    "\n",
    "for i, row in check_fastas.iterrows():\n",
    "\n",
    "    internal_id = row['Sample Name']\n",
    "    run_id = row['Run']\n",
    "\n",
    "    if not os.path.isfile(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_noID_1.fastq\"):\n",
    "        remove_headers_FASTQ(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_1.fastq\")\n",
    "\n",
    "    if not os.path.isfile(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_noID_2.fastq\"):\n",
    "        remove_headers_FASTQ(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_2.fastq\")\n",
    "        \n",
    "    if not os.path.isfile(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_noID_1.fastq\"):\n",
    "        remove_headers_FASTQ(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_1.fastq\")\n",
    "\n",
    "    if not os.path.isfile(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_noID_2.fastq\"):\n",
    "        remove_headers_FASTQ(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_2.fastq\")\n",
    "\n",
    "    cmp_1 = compare_files(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_noID_1.fastq\", f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_noID_1.fastq\")\n",
    "\n",
    "    cmp_2 = compare_files(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_noID_2.fastq\", f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_noID_2.fastq\")\n",
    "    \n",
    "    df_compare_FASTQ.loc[i, :] = [internal_id, cmp_1, cmp_2]\n",
    "    \n",
    "    if i % 50 == 0:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "314633eb-bf55-4b72-a52f-b7a5fac7b441",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for fName in glob.glob(\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/*/*_noID*\"):\n",
    "#     os.remove(fName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "efe1ed8b-db5a-403c-b0ed-ffd07b3fe692",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for fName in glob.glob(\"/n/scratch/users/s/sak0914/smk_output/*/*/*_noID*\"):\n",
    "#     os.remove(fName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0f0fd38-8512-4f6b-aa76-38c0868509e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bb84d1e3-91d0-4d33-b50e-ec90c661b653",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Run</th>\n",
       "      <th>FASTQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN13813815</td>\n",
       "      <td>SRR10868963</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN13813814</td>\n",
       "      <td>SRR10868964</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN13813813</td>\n",
       "      <td>SRR10868965</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN13813812</td>\n",
       "      <td>SRR10868966</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN13813811</td>\n",
       "      <td>SRR10868967</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>SAMN13813817</td>\n",
       "      <td>SRR10869319</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>SAMN13813816</td>\n",
       "      <td>SRR10869320</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>SAMN13813726</td>\n",
       "      <td>SRR10869321</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>SAMN13813717</td>\n",
       "      <td>SRR10869322</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>SAMN13813716</td>\n",
       "      <td>SRR10869323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        BioSample          Run  FASTQ\n",
       "0    SAMN13813815  SRR10868963      1\n",
       "1    SAMN13813814  SRR10868964      1\n",
       "2    SAMN13813813  SRR10868965      1\n",
       "3    SAMN13813812  SRR10868966      1\n",
       "4    SAMN13813811  SRR10868967      1\n",
       "..            ...          ...    ...\n",
       "356  SAMN13813817  SRR10869319      1\n",
       "357  SAMN13813816  SRR10869320      1\n",
       "358  SAMN13813726  SRR10869321      1\n",
       "359  SAMN13813717  SRR10869322      1\n",
       "360  SAMN13813716  SRR10869323      1\n",
       "\n",
       "[361 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_mapping_table_for_megapipe = cass_mapping_table[['BioSample', 'Run']]\n",
    "cass_mapping_table_for_megapipe['FASTQ'] = 1\n",
    "cass_mapping_table_for_megapipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f5dc0d-6062-4ed4-8739-cb863118cbfb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "125e570b-1a62-4b5e-9011-4eca950a2f55",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_fasta_sequences(file1, file2):\n",
    "    # Read sequences from both files into sets\n",
    "    sequences1 = {str(record.seq) for record in SeqIO.parse(file1, \"fasta\")}\n",
    "    sequences2 = {str(record.seq) for record in SeqIO.parse(file2, \"fasta\")}\n",
    "    \n",
    "    # Compare the sets of sequences\n",
    "    if sequences1 == sequences2:\n",
    "        return 0\n",
    "        \n",
    "    return 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f90c2cd7-adde-405c-b7b2-a36796411df0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kraken_db/PGAP/S0085-01.HybridAsm.PGAP.Genome.fasta kraken_db/Bakta/S0085-01.HybridAsm.Bakta.Genome.fasta\n",
      "kraken_db/PGAP/S0107-01.HybridAsm.PGAP.Genome.fasta kraken_db/Bakta/S0107-01.HybridAsm.Bakta.Genome.fasta\n"
     ]
    }
   ],
   "source": [
    "diff_fastas = {}\n",
    "alns_lst = []\n",
    "\n",
    "for fName in glob.glob(\"kraken_db/PGAP/*\"):\n",
    "    sample_ID = os.path.basename(fName).split('.')[0]\n",
    "\n",
    "    fName_2 = fName.replace('PGAP', 'Bakta')\n",
    "    assert os.path.isfile(fName_2)\n",
    "\n",
    "    return_val = compare_fasta_sequences(fName, fName_2)\n",
    "\n",
    "    if return_val == 1:\n",
    "        print(fName, fName_2)\n",
    "    \n",
    "    diff_fastas[sample_ID] = return_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "061dbf8d-2544-4fea-807b-d7ee803d3c21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(151, 151)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(os.listdir(\"kraken_db/PGAP\")), len(os.listdir(\"kraken_db/Bakta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c0ca6185-283b-48a0-a364-9bd6d6c6b993",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(151, 2)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_fastas_df = pd.DataFrame(diff_fastas, index=[0]).T.reset_index()\n",
    "diff_fastas_df.columns = ['sample', 'seq_diff']\n",
    "diff_fastas_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e79fe258-1e43-4961-998b-97fb97eea81b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>S0085-01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>S0107-01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample  seq_diff\n",
       "85  S0085-01         1\n",
       "88  S0107-01         1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_fastas_df.query(\"seq_diff==1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4b3bd043-1e40-471c-a5a4-2ea8d29b6744",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_id = 'S0085-01'\n",
    "seq_PGAP = [(seq.id, seq.seq) for seq in SeqIO.parse(f\"kraken_db/PGAP/{sample_id}.HybridAsm.PGAP.Genome.fasta\", \"fasta\")]\n",
    "seq_Bakta = [(seq.id, seq.seq) for seq in SeqIO.parse(f\"kraken_db/Bakta/{sample_id}.HybridAsm.Bakta.Genome.fasta\", \"fasta\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7d5b80a8-5e45-45a4-b770-e1d48334c9f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4417316, 4417384)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seq_PGAP[0][1]), len(seq_Bakta[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7a7b243d-1d81-4a6f-8d84-387036a6a206",
   "metadata": {},
   "outputs": [],
   "source": [
    "S0085_check = \"/n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/TRUST_PB_Set1/S0085-01.LR.Asm.fasta\"\n",
    "S0085_check_fastq = [(seq.id, seq.seq) for seq in SeqIO.parse(S0085_check, \"fasta\")]\n",
    "\n",
    "S0107_check = \"/n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/TRUST_PB_Set1/S0107-01.LR.Asm.fasta\"\n",
    "S0107_check_fastq = [(seq.id, seq.seq) for seq in SeqIO.parse(S0107_check, \"fasta\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7bdd830f-457a-4751-b3c1-241eaf9afd68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4417384, 4416462)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(S0085_check_fastq[0][1]), len(S0107_check_fastq[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2e43e952-7228-4e0c-afb9-5f2e1a0c55fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_id = 'S0107-01'\n",
    "seq_PGAP = [(seq.id, seq.seq) for seq in SeqIO.parse(f\"kraken_db/PGAP/{sample_id}.HybridAsm.PGAP.Genome.fasta\", \"fasta\")]\n",
    "seq_Bakta = [(seq.id, seq.seq) for seq in SeqIO.parse(f\"kraken_db/Bakta/{sample_id}.HybridAsm.Bakta.Genome.fasta\", \"fasta\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ed996fc0-5235-4b58-99b0-80fefe78ca3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4416405, 4416462)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seq_PGAP[0][1]), len(seq_Bakta[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2496d0a6-c2e6-4fbe-b2cf-2c73ad281a18",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "13698c04-ae40-461e-bf0e-c8442835b903",
   "metadata": {},
   "source": [
    "# Create BED file to exclude sites with low EBR (refined low confidence) or are within 50 bp of a PE/PPE gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c1138b10-acd0-43ad-be9e-2e151f750d40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude_fName_Marin = 'RLC_Regions.Plus.LowPmapK50E4.H37Rv'\n",
    "\n",
    "# RLC_regions = pd.read_csv(f\"references/ref_genome/{exclude_fName_Marin}.bed\", sep='\\t', header=None)\n",
    "# RLC_regions.columns = ['CHR', 'BEG', 'END']\n",
    "# RLC_regions['CHR'] = 'Chromosome'\n",
    "# len(RLC_regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "bc89c459-1f0c-4129-aac6-e2534da7b32c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19924\n"
     ]
    }
   ],
   "source": [
    "# exclude PE/PPE/PGRS genes, sites with EBR < 0.99, and homoplasic sites from Vargas et al., PNAS 2023 \n",
    "EBR_fName_Marin = 'EBR_Marin.bedgraph'\n",
    "\n",
    "df_EBR = pd.read_csv(f\"references/ref_genome/{EBR_fName_Marin}\", sep='\\t', header=None)\n",
    "df_EBR.columns = ['CHR', 'BEG', 'END', 'EBR']\n",
    "df_EBR['CHR'] = 'Chromosome'\n",
    "\n",
    "exclude_genes = H37Rv_regions.query(\"Name.str.startswith('PE') | Name.str.startswith('PPE') | Name.str.contains('PGRS')\").Name.values\n",
    "\n",
    "PE_PPE_BED_file_df = pd.DataFrame(columns=['CHR', 'BEG', 'END'])\n",
    "\n",
    "for i, gene in enumerate(exclude_genes):\n",
    "\n",
    "    # these are 1-indexed inclusive\n",
    "    gene_start, gene_end = H37Rv_regions.set_index('Name').loc[gene, ['Start', 'Stop']].values\n",
    "\n",
    "    # BED file is 0-indexed half open interval\n",
    "    # also extend the range by 50 base pairs on either side\n",
    "    gene_start = gene_start - 1 - 50\n",
    "    gene_end = gene_end + 50\n",
    "\n",
    "    PE_PPE_BED_file_df.loc[i, :] = ['Chromosome', gene_start, gene_end]\n",
    "\n",
    "homoplasy = pd.read_excel(\"./references/phylogeny/Vargas_PNAS_2023_homoplasy.xlsx\", sheet_name=None)\n",
    "\n",
    "if len(homoplasy.keys()) == 1:\n",
    "    homoplasy = homoplasy['Sheet1']\n",
    "\n",
    "homoplasic_sites_BED_file = pd.DataFrame(columns=['CHR', 'BEG', 'END'])\n",
    "\n",
    "for i, row in homoplasy.iterrows():\n",
    "    homoplasic_sites_BED_file.loc[i, :] = ['Chromosome', row['H37Rv Position']-1, row['H37Rv Position']]\n",
    "\n",
    "BED_file_df = pd.concat([PE_PPE_BED_file_df, df_EBR.query(\"EBR < 0.99\")[['CHR', 'BEG', 'END']], homoplasic_sites_BED_file]).sort_values([\"BEG\", \"END\"], ascending=True)\n",
    "print(len(BED_file_df))\n",
    "\n",
    "BED_file_df.sort_values(['BEG', 'END']).to_csv(\"./references/phylogeny/exclude_regions_pre_merge.bed\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "d4297029-fe33-4cf6-aeba-4ae3f66a96c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19924\n"
     ]
    }
   ],
   "source": [
    "bedtools merge -i /home/sak0914/Mtb_Megapipe/references/phylogeny/exclude_regions_pre_merge.bed > /home/sak0914/Mtb_Megapipe/references/phylogeny/exclude_regions.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92ca7c5d-a3af-45b8-ba87-4a321b00c4f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gunzip -c SAMEA3392614_full.vcf.gz | bcftools filter -e \"SVTYPE == 'INS' | SVTYPE == 'DEL' | IC > 0 | DC > 0\" | bedtools subtract -a '-' -b '/home/sak0914/Mtb_Megapipe/references/phylogeny/exclude_regions_combined.bed' | SnpSift extractFields '-' POS REF ALT FILTER QUAL AF DP BQ MQ -e \"\" >  /n/scratch/users/s/sak0914/smk_output/SAMEA3392614/lineage/SNP_sites.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "29d0b62e-5348-4ed3-b7d0-379b688d5964",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>174</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>188</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>212</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>274</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3770</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4408093</td>\n",
       "      <td>4408094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3771</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4408919</td>\n",
       "      <td>4408920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3772</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4411370</td>\n",
       "      <td>4411400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3773</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4411469</td>\n",
       "      <td>4411471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3774</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4411474</td>\n",
       "      <td>4411475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3775 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1        2\n",
       "0     Chromosome        0       33\n",
       "1     Chromosome      174      185\n",
       "2     Chromosome      188      211\n",
       "3     Chromosome      212      244\n",
       "4     Chromosome      274      275\n",
       "...          ...      ...      ...\n",
       "3770  Chromosome  4408093  4408094\n",
       "3771  Chromosome  4408919  4408920\n",
       "3772  Chromosome  4411370  4411400\n",
       "3773  Chromosome  4411469  4411471\n",
       "3774  Chromosome  4411474  4411475\n",
       "\n",
       "[3775 rows x 3 columns]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"/home/sak0914/Mtb_Megapipe/references/phylogeny/exclude_regions.bed\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ca1abdd-e7f4-4461-ba55-df293f9c9b0d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e49414b-d8ea-4544-8993-77fb637e21d5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "0498432b-ca33-455f-a4ea-9d0994294f1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_SNP_matrix_from_VCF(isolate, qual_thresh=10, min_read_depth=5, BQ_thresh=20, MQ_thresh=30):\n",
    "\n",
    "    # for encoding ambiguous alleles as possibly both nucleotides\n",
    "    iupac_two_nuc_codes = {\n",
    "                            'AG': 'R',  # Purine (A or G)\n",
    "                            'CT': 'Y',  # Pyrimidine (C or T)\n",
    "                            'CG': 'S',  # Strong (C or G)\n",
    "                            'AT': 'W',  # Weak (A or T)\n",
    "                            'GT': 'K',  # Keto (G or T)\n",
    "                            'AC': 'M',  # Amino (A or C),\n",
    "                            'AA': 'A',  # the following are so that you don't get a key error, but they don't actually need to be merged\n",
    "                            'CC': 'C',\n",
    "                            'GG': 'G',\n",
    "                            'TT': 'T'\n",
    "                          }\n",
    "    \n",
    "    # keep only SNPs (non-length changing variants)\n",
    "    SNP_sites = pd.read_csv(f\"/n/scratch/users/s/sak0914/smk_output/{isolate}/lineage/SNP_sites.tsv.gz\", sep='\\t', compression='gzip').query(\"REF.str.len() == ALT.str.len()\").drop_duplicates().reset_index(drop=True)\n",
    "    \n",
    "    # these are REF alleles\n",
    "    SNP_sites.loc[SNP_sites['ALT']=='.', 'ALT'] = SNP_sites['REF']\n",
    "\n",
    "    # Reference\n",
    "    SNP_sites.loc[(SNP_sites['FILTER'].isin(['PASS', 'Amb'])) &\n",
    "                  (SNP_sites['QUAL'] >= qual_thresh) &\n",
    "                  (SNP_sites['DP'] >= min_read_depth) &\n",
    "                  (SNP_sites['BQ'] >= BQ_thresh) &\n",
    "                  (SNP_sites['MQ'] >= MQ_thresh) &\n",
    "                  (SNP_sites['AF'] <= 0.25), 'ALLELE'] = SNP_sites['REF']\n",
    "    \n",
    "    # high quality variants\n",
    "    SNP_sites.loc[(SNP_sites['FILTER']=='PASS') &\n",
    "                  (SNP_sites['QUAL'] >= qual_thresh) &\n",
    "                  (SNP_sites['DP'] >= min_read_depth) &\n",
    "                  (SNP_sites['BQ'] >= BQ_thresh) &\n",
    "                  (SNP_sites['MQ'] >= MQ_thresh) &\n",
    "                  (SNP_sites['AF'] > 0.75), 'ALLELE'] = SNP_sites['ALT']\n",
    "    \n",
    "    # ambiguous variants -- use the IUPAC codes representing the two nucleotides (REF and ALT) at a given site\n",
    "    # faster to not apply it to all rows, only on the ones you need\n",
    "    SNP_sites.loc[(SNP_sites['FILTER']=='Amb') &\n",
    "                  (SNP_sites['QUAL'] >= qual_thresh) &\n",
    "                  (SNP_sites['DP'] >= min_read_depth) &\n",
    "                  (SNP_sites['BQ'] >= BQ_thresh) &\n",
    "                  (SNP_sites['MQ'] >= MQ_thresh) &\n",
    "                  (SNP_sites['AF'] > 0.25), 'ALLELE'] = SNP_sites.loc[(SNP_sites['FILTER']=='Amb') &\n",
    "                                                                      (SNP_sites['QUAL'] >= qual_thresh) &\n",
    "                                                                      (SNP_sites['DP'] >= min_read_depth) &\n",
    "                                                                      (SNP_sites['BQ'] >= BQ_thresh) &\n",
    "                                                                      (SNP_sites['MQ'] >= MQ_thresh) &\n",
    "                                                                      (SNP_sites['AF'] > 0.25)].apply(lambda row: iupac_two_nuc_codes.get(''.join(sorted(row['REF'] + row['ALT'])), np.nan), axis=1)\n",
    "    \n",
    "    # fill the rest with N = no information\n",
    "    # this includes low coverage sites because we don't know what's happening there, so in the absence of other information, it could be anything\n",
    "    # also includes areas with deletions -- those shouldn't be REF, technically they are '-', which is treated the same as N\n",
    "    # gap characters ('-') and Ns are treated the same in iqtree\n",
    "    SNP_sites['ALLELE'] = SNP_sites['ALLELE'].fillna('N')\n",
    "    \n",
    "    SNP_sites['ISOLATE'] = isolate\n",
    "\n",
    "    # return a pivoted matrix\n",
    "    return SNP_sites.pivot(index='ISOLATE', columns='POS', values='ALLELE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "8dc73cdd-4282-4be0-b68d-ec7cfeafa763",
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_matrix_full = []\n",
    "\n",
    "for isolate in os.listdir(\"/n/scratch/users/s/sak0914/smk_output\"):\n",
    "\n",
    "    SNP_fName = f\"/n/scratch/users/s/sak0914/smk_output/{isolate}/lineage/SNP_sites.tsv.gz\"\n",
    "    \n",
    "    if os.path.isfile(SNP_fName):\n",
    "        SNP_matrix = create_SNP_matrix_from_VCF(isolate)\n",
    "        SNP_matrix_full.append(SNP_matrix)\n",
    "\n",
    "SNP_matrix_full = pd.concat(SNP_matrix_full)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "1859fc42-29da-4b9a-affb-c21d512cd448",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>POS</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>36</th>\n",
       "      <th>37</th>\n",
       "      <th>38</th>\n",
       "      <th>39</th>\n",
       "      <th>40</th>\n",
       "      <th>41</th>\n",
       "      <th>42</th>\n",
       "      <th>43</th>\n",
       "      <th>...</th>\n",
       "      <th>4411523</th>\n",
       "      <th>4411524</th>\n",
       "      <th>4411525</th>\n",
       "      <th>4411526</th>\n",
       "      <th>4411527</th>\n",
       "      <th>4411528</th>\n",
       "      <th>4411529</th>\n",
       "      <th>4411530</th>\n",
       "      <th>4411531</th>\n",
       "      <th>4411532</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISOLATE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>00-R1210</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3392614</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA8741574</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 3956879 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "POS          34      35      36      37      38      39      40      41       \\\n",
       "ISOLATE                                                                        \n",
       "00-R1210           G       T       G       T       G       G       A       A   \n",
       "SAMEA3392614       G       T       G       T       G       G       A       A   \n",
       "SAMEA8741574       G       T       G       T       G       G       A       A   \n",
       "\n",
       "POS          42      43       ... 4411523 4411524 4411525 4411526 4411527  \\\n",
       "ISOLATE                       ...                                           \n",
       "00-R1210           C       G  ...       A       G       A       T       A   \n",
       "SAMEA3392614       C       G  ...       A       G       A       T       A   \n",
       "SAMEA8741574       C       G  ...       A       G       A       T       A   \n",
       "\n",
       "POS          4411528 4411529 4411530 4411531 4411532  \n",
       "ISOLATE                                               \n",
       "00-R1210           C       G       T       C       G  \n",
       "SAMEA3392614       C       G       T       C       G  \n",
       "SAMEA8741574       C       G       T       C       G  \n",
       "\n",
       "[3 rows x 3956879 columns]"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_matrix_full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "22030dac-d375-4427-99e4-0adddac40f78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3, 56550)"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove sites that are the same across all isolates because they have no signal\n",
    "SNP_matrix_no_constant = SNP_matrix_full.loc[:, SNP_matrix_full.nunique(dropna=False) > 1]\n",
    "SNP_matrix_no_constant.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "ed86286c-aae6-4d20-be4a-ec40c13b66a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>POS</th>\n",
       "      <th>1452</th>\n",
       "      <th>1849</th>\n",
       "      <th>4348</th>\n",
       "      <th>4349</th>\n",
       "      <th>4350</th>\n",
       "      <th>4351</th>\n",
       "      <th>4352</th>\n",
       "      <th>4353</th>\n",
       "      <th>4354</th>\n",
       "      <th>4355</th>\n",
       "      <th>...</th>\n",
       "      <th>4397736</th>\n",
       "      <th>4400246</th>\n",
       "      <th>4401732</th>\n",
       "      <th>4404346</th>\n",
       "      <th>4404960</th>\n",
       "      <th>4406116</th>\n",
       "      <th>4407200</th>\n",
       "      <th>4407588</th>\n",
       "      <th>4407927</th>\n",
       "      <th>4408923</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISOLATE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>00-R1210</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3392614</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA8741574</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 56550 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "POS          1452    1849    4348    4349    4350    4351    4352    4353     \\\n",
       "ISOLATE                                                                        \n",
       "00-R1210           T       C       N       N       N       N       N       N   \n",
       "SAMEA3392614       G       A       G       A       G       G       A       T   \n",
       "SAMEA8741574       G       C       G       A       G       G       A       T   \n",
       "\n",
       "POS          4354    4355     ... 4397736 4400246 4401732 4404346 4404960  \\\n",
       "ISOLATE                       ...                                           \n",
       "00-R1210           N       N  ...       C       A       C       T       C   \n",
       "SAMEA3392614       A       T  ...       T       G       A       C       C   \n",
       "SAMEA8741574       A       T  ...       C       G       C       C       G   \n",
       "\n",
       "POS          4406116 4407200 4407588 4407927 4408923  \n",
       "ISOLATE                                               \n",
       "00-R1210           C       G       T       T       C  \n",
       "SAMEA3392614       C       A       C       G       T  \n",
       "SAMEA8741574       A       A       T       T       C  \n",
       "\n",
       "[3 rows x 56550 columns]"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_matrix_no_constant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21083fd4-48a6-4cbe-9ee5-4a943f2e5426",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove sites where more than 10% of isolates have Ns -- this is a param that can be changed by the user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81bd624d-196c-451b-96b4-cd9a8ba4093f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "44d79928-8527-4154-8b69-807d8658aac0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'kraken_db/BAKTA/01_R1134.HybridAsm.Bakta.Genome.fasta'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob(\"kraken_db/BAKTA/*\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f7874046-bbc7-4fe1-beb2-d30b61fae314",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'kraken_db/PGAP/01_R1134.HybridAsm.PGAP.Genome.fasta'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob(\"kraken_db/PGAP/*\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7fa2b110-842a-4e6a-9314-26d7c07994cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'kraken_db/BAKTA/01_R1134.HybridAsm.BAKTA.Genome.fasta'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fName.replace('PGAP', 'BAKTA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c74ee7fa-c3e7-4092-8c7c-9b8c05f49c9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sequences in the two FASTA files are the same.\n"
     ]
    }
   ],
   "source": [
    "# Example usage\n",
    "compare_fasta_sequences(\"kraken_db/01_R1134.HybridAsm.Bakta.Genome.fasta\", \"kraken_db/01_R1134.HybridAsm.PGAP.Genome.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3088982c-4713-4bda-aee8-622f88daa2f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4319, 1)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "todo_isolates = pd.read_csv(\"../data_cleaning/samples_for_F2.txt\", header=None, sep='\\t')\n",
    "todo_isolates.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3161ae9a-13e7-435e-b724-09354e2d9719",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3863</th>\n",
       "      <td>SAMN37457963</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0\n",
       "3863  SAMN37457963"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "todo_isolates.loc[todo_isolates[0]=='SAMN37457963']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1884bc8e-a2e0-4eb6-9182-40b13bd36890",
   "metadata": {},
   "outputs": [],
   "source": [
    "python3 MtbQuantCNN/data_processing/variant_calling/F2_calculation.py data_cleaning/samples_for_F2.txt /n/data1/hms/dbmi/farhat/rollingDB/genomic_data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
